From ae682cb8e4d3e5fbafd96c3e23ad566390e250c6 Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Thu, 14 Aug 2025 11:43:02 +0800
Subject: [PATCH] comctl32/combo: Don't draw a focus rect when keyboard cues
 are disabled.

---
 dlls/comctl32/combo.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/comctl32/combo.c b/dlls/comctl32/combo.c
index 1094132826b..a984adc0d5d 100644
--- a/dlls/comctl32/combo.c
+++ b/dlls/comctl32/combo.c
@@ -716,7 +716,7 @@ static void CBPaintText(HEADCOMBO *lphc, HDC hdc_paint)
 		    &rectEdit,
 		    pText ? pText : L"" , size, NULL );
 
-       if(lphc->wState & CBF_FOCUSED && !(lphc->wState & CBF_DROPPED))
+       if(COMCTL32_keyboard_cues_enabled && lphc->wState & CBF_FOCUSED && !(lphc->wState & CBF_DROPPED))
 	 DrawFocusRect( hdc, &rectEdit );
      }
 
-- 
GitLab

