From f7ae87cd33243dd881ad5e3d6776a18126376175 Mon Sep 17 00:00:00 2001
From: Conor McCarthy <cmccarthy@codeweavers.com>
Date: Mon, 3 Mar 2025 12:21:21 +1000
Subject: [PATCH] mf: Move transform node reset to a helper function.

---
 dlls/mf/session.c | 32 +++++++++++++++++++-------------
 1 file changed, 19 insertions(+), 13 deletions(-)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index 483ea6f904f..e595cadd185 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -1061,13 +1061,29 @@ static void session_handle_start_error(struct media_session *session, HRESULT hr
     session_command_complete_with_event(session, MESessionStarted, hr, NULL);
 }
 
+static void session_reset_transforms(struct media_session *session)
+{
+    struct topo_node *topo_node;
+    UINT i;
+
+    LIST_FOR_EACH_ENTRY(topo_node, &session->presentation.nodes, struct topo_node, entry)
+    {
+        if (topo_node->type != MF_TOPOLOGY_TRANSFORM_NODE)
+            continue;
+
+        for (i = 0; i < topo_node->u.transform.input_count; i++)
+        {
+            struct transform_stream *stream = &topo_node->u.transform.inputs[i];
+            stream->draining = FALSE;
+        }
+    }
+}
+
 static void session_start(struct media_session *session, const GUID *time_format, const PROPVARIANT *start_position)
 {
     struct media_source *source;
-    struct topo_node *topo_node;
     MFTIME duration;
     HRESULT hr;
-    UINT i;
 
     switch (session->state)
     {
@@ -1103,17 +1119,7 @@ static void session_start(struct media_session *session, const GUID *time_format
                 }
             }
 
-            LIST_FOR_EACH_ENTRY(topo_node, &session->presentation.nodes, struct topo_node, entry)
-            {
-                if (topo_node->type == MF_TOPOLOGY_TRANSFORM_NODE)
-                {
-                    for (i = 0; i < topo_node->u.transform.input_count; i++)
-                    {
-                        struct transform_stream *stream = &topo_node->u.transform.inputs[i];
-                        stream->draining = FALSE;
-                    }
-                }
-            }
+            session_reset_transforms(session);
 
             session->state = SESSION_STATE_STARTING_SOURCES;
             break;
-- 
GitLab

