From cc6970b7ec5c3e9599a77a6356dbe5715aadd393 Mon Sep 17 00:00:00 2001
From: Conor McCarthy <cmccarthy@codeweavers.com>
Date: Fri, 28 Feb 2025 14:08:49 +1000
Subject: [PATCH] mf: Reset transform nodes when seeking.

---
 dlls/mf/session.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/mf/session.c b/dlls/mf/session.c
index 7e68c89212e..8a55a97b13a 100644
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -1163,6 +1163,8 @@ static void session_start(struct media_session *session, const GUID *time_format
                 }
             }
 
+            session_reset_transforms(session, TRUE);
+
             session->presentation.time_format = *time_format;
             session->presentation.start_position.vt = VT_EMPTY;
             PropVariantCopy(&session->presentation.start_position, start_position);
-- 
GitLab

