From 677e9ed63dd459ef0104abfb9f6190c198806750 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sat, 14 Jun 2025 00:15:34 -0700
Subject: [PATCH] winex11: Hotfix an issue on nvidia for this patchset.

---
 dlls/winex11.drv/opengl.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winex11.drv/opengl.c b/dlls/winex11.drv/opengl.c
index e441b538009..174865f165d 100644
--- a/dlls/winex11.drv/opengl.c
+++ b/dlls/winex11.drv/opengl.c
@@ -1786,7 +1786,7 @@ static BOOL x11drv_swap_buffers( void *private, HWND hwnd, HDC hdc, int interval
     set_swap_interval( gl, interval );
     pthread_mutex_unlock( &context_mutex );
 
-    if (!ctx || !gl->offscreen) pglXSwapBuffers( gdi_display, gl->drawable );
+    if (!ctx || !gl->offscreen || !pglXSwapBuffersMscOML) pglXSwapBuffers( gdi_display, gl->drawable );
     else
     {
         funcs->p_glFlush();
-- 
2.49.0

