From 0beeef66195534429d1d9bed20aa88143c1f2042 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 20 Nov 2025 21:48:25 -0800
Subject: [PATCH] include: Allow force-disabling compiler exceptions.

---
 include/excpt.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/excpt.h b/include/excpt.h
index 17c7409c636..59fc2475abd 100644
--- a/include/excpt.h
+++ b/include/excpt.h
@@ -38,7 +38,7 @@ typedef enum _EXCEPTION_DISPOSITION
 #define EXCEPTION_CONTINUE_EXECUTION    -1
 
 
-#if !defined(USE_COMPILER_EXCEPTIONS) && defined(_MSC_VER) && !defined(__i386__) && !defined(__arm__) && (!defined(__clang_major__) || __clang_major__ >= 19)
+#if !defined(NO_COMPILER_EXCEPTIONS) && !defined(USE_COMPILER_EXCEPTIONS) && defined(_MSC_VER) && !defined(__i386__) && !defined(__arm__) && (!defined(__clang_major__) || __clang_major__ >= 19)
 #define USE_COMPILER_EXCEPTIONS
 #endif
 
-- 
2.51.2

