diff --git a/libs/musl/Makefile.in b/libs/musl/Makefile.in
index bcb14ffef78..e3f4461360e 100644
--- a/libs/musl/Makefile.in
+++ b/libs/musl/Makefile.in
@@ -1,6 +1,6 @@
 EXTLIB    = libmusl.a
 EXTRAINCL = -I$(srcdir)/src/internal -I$(srcdir)/arch/generic
-EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE
+EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE -fexcess-precision=standard
 
 SOURCES = \
	 src/math/__cos.c \
