diff --git a/libs/musl/Makefile.in b/libs/musl/Makefile.in
index bcb14ffef78..e3f4461360e 100644
--- a/libs/musl/Makefile.in
+++ b/libs/musl/Makefile.in
@@ -1,6 +1,6 @@
 EXTLIB    = libmusl.a
 EXTRAINCL = -I$(srcdir)/src/internal -I$(srcdir)/arch/generic
-EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE
+EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE -fexcess-precision=standard -std=c11 -ffloat-store
 
 SOURCES = \
	 src/math/__cos.c \
diff --git a/libs/musl/src/internal/libm.h b/libs/musl/src/internal/libm.h
index a1e9bc08716..5621f9dec39 100644
--- a/libs/musl/src/internal/libm.h
+++ b/libs/musl/src/internal/libm.h
@@ -7,6 +7,12 @@
 #include <errno.h>
 #include <features.h>
 
+#define M_PI		3.14159265358979323846	/* pi */
+#define M_PI_2		1.57079632679489661923	/* pi/2 */
+#define M_PI_4		0.78539816339744830962	/* pi/4 */
+
+//#pragma GCC target ("arch=x86-64","fpmath=387","no-sse")
+
 typedef float float_t;
 typedef double double_t;
 
