From d43631c794412029bfd804349a081be9395ab313 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sun, 29 Dec 2024 23:23:41 -0800
Subject: [PATCH] configure, ntdll: Use gettid() if available.

---
 configure.ac             | 1 +
 dlls/ntdll/unix/server.c | 2 ++
 2 files changed, 3 insertions(+)

diff --git a/configure.ac b/configure.ac
index 10f78f7af86..6d0562d50b9 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2089,6 +2089,7 @@ AC_CHECK_FUNCS(\
 	getauxval \
 	getifaddrs \
 	getrandom \
+ 	gettid \
 	kqueue \
 	pipe2 \
 	port_create \
diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index 5862cd1666f..d9452e168ab 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -1549,6 +1549,8 @@ static int get_unix_tid(void)
     int ret = -1;
 #ifdef HAVE_PTHREAD_GETTHREADID_NP
     ret = pthread_getthreadid_np();
+#elif defined(HAVE_GETTID)
+    ret = gettid();
 #elif defined(linux)
     ret = syscall( __NR_gettid );
 #elif defined(__sun)
-- 
2.47.1

