From 83b5d925d494479b526532e70a66170223c86ccf Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Fri, 31 Jan 2025 13:46:42 -0800
Subject: [PATCH] msvcrt: Allow SSE if _MSVCR_VER == 0.

---
 dlls/msvcrt/math.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/msvcrt/math.c b/dlls/msvcrt/math.c
index bcdf459396b..ceffe564cea 100644
--- a/dlls/msvcrt/math.c
+++ b/dlls/msvcrt/math.c
@@ -70,7 +70,7 @@ static BOOL sse2_enabled;
 void msvcrt_init_math( void *module )
 {
     sse2_supported = IsProcessorFeaturePresent( PF_XMMI64_INSTRUCTIONS_AVAILABLE );
-#if _MSVCR_VER <=71
+#if _MSVCR_VER > 0 && _MSVCR_VER <=71
     sse2_enabled = FALSE;
 #else
     sse2_enabled = sse2_supported;
-- 
2.48.1

