From 695afb8b670f44209200636be7c3581a18ba4c3f Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Fri, 21 Feb 2025 01:01:15 +0900
Subject: [PATCH] ntdll: Remove superflous NULL check for importer.

---
 dlls/ntdll/loader.c | 11 ++++-------
 1 file changed, 4 insertions(+), 7 deletions(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 7a8bbc77aef..144f0bb2096 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -739,7 +739,7 @@ static NTSTATUS build_import_name( WINE_MODREF *importer, WCHAR buffer[256], con
 {
     const API_SET_NAMESPACE *map = NtCurrentTeb()->Peb->ApiSetMap;
     const API_SET_NAMESPACE_ENTRY *entry;
-    const WCHAR *host = importer ? importer->ldr.BaseDllName.Buffer : NULL;
+    const WCHAR *host = importer->ldr.BaseDllName.Buffer;
     UNICODE_STRING str;
 
     while (len && import[len-1] == ' ') len--;  /* remove trailing spaces */
@@ -956,12 +956,9 @@ static FARPROC find_forwarded_export( HMODULE module, const char *forward, LPCWS
 
     if (wm->ldr.DdagNode != node_ntdll && wm->ldr.DdagNode != node_kernel32)
     {
-        if (importer)
-        {
-            /* Prepare for the callee stealing the reference */
-            if (!wm_loaded && wm->ldr.LoadCount != -1) wm->ldr.LoadCount++;
-            add_module_dependency( importer->ldr.DdagNode, wm->ldr.DdagNode );
-        }
+        /* Prepare for the callee stealing the reference */
+        if (!wm_loaded && wm->ldr.LoadCount != -1) wm->ldr.LoadCount++;
+        add_module_dependency( importer->ldr.DdagNode, wm->ldr.DdagNode );
         if (is_dynamic && wm_loaded && process_attach( wm->ldr.DdagNode, NULL ) != STATUS_SUCCESS)
         {
             ERR( "process_attach failed for forward '%s' used by %s\n",
-- 
GitLab

