From 44c2e13466558ae762955005fb778a52719bbe41 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 18 Feb 2025 10:50:03 +0100
Subject: [PATCH] explorer: Paint the desktop even without RDW_ERASE.

---
 programs/explorer/desktop.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/explorer/desktop.c b/programs/explorer/desktop.c
index e0b57a4b37f..ce70c5dd79d 100644
--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -842,7 +842,7 @@ static LRESULT WINAPI desktop_wnd_proc( HWND hwnd, UINT message, WPARAM wp, LPAR
             BeginPaint( hwnd, &ps );
             if (!using_root)
             {
-                if (ps.fErase) PaintDesktop( ps.hdc );
+                PaintDesktop( ps.hdc );
                 draw_launchers( ps.hdc, ps.rcPaint );
             }
             EndPaint( hwnd, &ps );
-- 
GitLab

