From df55bcc4c1469aeaaeee8272ef6806e35813922e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Wed, 19 Mar 2025 11:43:58 +0100
Subject: [PATCH] hidclass: Dispatch IRP_MN_SURPRISE_REMOVAL to the
 minidrivers.

---
 dlls/hidclass.sys/pnp.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/hidclass.sys/pnp.c b/dlls/hidclass.sys/pnp.c
index c8d11cf43ba..aa58b8e288c 100644
--- a/dlls/hidclass.sys/pnp.c
+++ b/dlls/hidclass.sys/pnp.c
@@ -416,6 +416,7 @@ static NTSTATUS fdo_pnp(DEVICE_OBJECT *device, IRP *irp)
             return status;
 
         case IRP_MN_SURPRISE_REMOVAL:
+            if ((status = minidriver->PNPDispatch( device, irp ))) return status;
             KeSetEvent( &fdo->halt_event, IO_NO_INCREMENT, FALSE );
             return STATUS_SUCCESS;
 
-- 
GitLab

