From 94e0cba5d1b6e1377ff677af9c434f671b5fa777 Mon Sep 17 00:00:00 2001
From:
Date: Tue, 25 Jun 2024 04:28:02 +0000
Subject: [PATCH (REBASE) 07/10] winex11: Create a window surface even when there is
 client window.

---
 dlls/winex11.drv/bitblt.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/winex11.drv/bitblt.c b/dlls/winex11.drv/bitblt.c
index c0ffa1531a0..894004b1a38 100644
--- a/dlls/winex11.drv/bitblt.c
+++ b/dlls/winex11.drv/bitblt.c
@@ -1962,7 +1962,6 @@ static BOOL enable_direct_drawing( struct x11drv_win_data *data, BOOL layered )
     if (layered) return FALSE;
     if (data->embedded) return TRUE; /* draw directly to the window */
     if (data->whole_window == root_window) return TRUE; /* draw directly to the window */
-    if (data->client_window) return TRUE; /* draw directly to the window */
     if (!client_side_graphics) return TRUE; /* draw directly to the window */
     return FALSE;
 }
