From 94e0cba5d1b6e1377ff677af9c434f671b5fa777 Mon Sep 17 00:00:00 2001
From:
Date: Tue, 25 Jun 2024 04:28:02 +0000
Subject: [PATCH (REBASE) 07/10] winex11: Create a window surface even when there is
 client window.

---
 dlls/winex11.drv/bitblt.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/winex11.drv/bitblt.c b/dlls/winex11.drv/bitblt.c
index 0db8087441a..d1219862f74 100644
--- a/dlls/winex11.drv/bitblt.c
+++ b/dlls/winex11.drv/bitblt.c
@@ -2263,7 +2263,6 @@ BOOL X11DRV_CreateWindowSurface( HWND hwnd, UINT swp_flags, const RECT *visible_
         *surface = NULL;  /* indicate that we want to draw directly to the window */
         if (data->embedded) goto done; /* draw directly to the window */
         if (data->whole_window == root_window) goto done; /* draw directly to the window */
-        if (data->client_window) goto done; /* draw directly to the window */
         if (!client_side_graphics) goto done; /* draw directly to the window */
     }
 
