diff --git a/dlls/winex11.drv/vulkan.c b/dlls/winex11.drv/vulkan.c
index c2f32f4047a..af4e02c36a9 100644
--- a/dlls/winex11.drv/vulkan.c
+++ b/dlls/winex11.drv/vulkan.c
@@ -113,7 +113,7 @@ static void X11DRV_vulkan_surface_attach( HWND hwnd, void *private )
     if ((data = get_win_data( hwnd )))
     {
 #ifdef SONAME_LIBXCOMPOSITE
-        if (usexcomposite) pXCompositeUnredirectWindow( gdi_display, client_window, CompositeRedirectManual );
+        if (usexcomposite && !gl_disable_child_window_rendering) pXCompositeUnredirectWindow( gdi_display, client_window, CompositeRedirectManual );
 #endif
         attach_client_window( data, client_window );
         release_win_data( data );
@@ -144,7 +144,7 @@ static void X11DRV_vulkan_surface_detach( HWND hwnd, void *private, HDC *hdc )
         NtUserGetClientRect( hwnd, &escape.dc_rect, get_win_monitor_dpi( hwnd ) );
         NtGdiExtEscape( *hdc, NULL, 0, X11DRV_ESCAPE, sizeof(escape), (LPSTR)&escape, 0, NULL );
 #ifdef SONAME_LIBXCOMPOSITE
-        if (usexcomposite) pXCompositeRedirectWindow( gdi_display, client_window, CompositeRedirectManual );
+        if (usexcomposite && !gl_disable_child_window_rendering) pXCompositeRedirectWindow( gdi_display, client_window, CompositeRedirectManual );
 #endif
     }
 }
diff --git a/dlls/win32u/vulkan.c b/dlls/win32u/vulkan.c
index 47c1ca1ef01..96cf3fdad26 100644
--- a/dlls/win32u/vulkan.c
+++ b/dlls/win32u/vulkan.c
@@ -46,6 +46,8 @@ void *(*p_vkGetInstanceProcAddr)(VkInstance, const char *) = NULL;
 static void *vulkan_handle;
 static struct vulkan_funcs vulkan_funcs;
 
+static BOOL gl_disable_child_window_rendering = TRUE;
+
 #ifdef SONAME_LIBVULKAN
 
 static const struct vulkan_driver_funcs *driver_funcs;
@@ -398,6 +400,9 @@ static void vulkan_init_once(void)
 
 BOOL vulkan_init(void)
 {
+    const char *e = getenv("WINE_DISABLE_GLCHILD_HACK");
+    if (e && *e != '\0' && *e != '0') gl_disable_child_window_rendering = FALSE;
+
     static pthread_once_t init_once = PTHREAD_ONCE_INIT;
     pthread_once( &init_once, vulkan_init_once );
     return !!vulkan_handle;
@@ -487,7 +492,7 @@ void vulkan_set_region( HWND toplevel, HRGN region )
         
         NtUserMapWindowPoints( surface->hwnd, toplevel, (POINT *)&client_rect, 2, 0 /* per-monitor DPI */ );
         NtUserGetClientRect( surface->hwnd, &client_rect, get_win_monitor_dpi(surface->hwnd) );
-        is_clipped = NtGdiRectInRegion( region, &client_rect );
+        is_clipped = !gl_disable_child_window_rendering && NtGdiRectInRegion( region, &client_rect ) ;
 
         if (is_clipped && !surface->offscreen_dc)
         {
