From 4b5f64b6ea2b8d4b0398c03a75882221c9da463d Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Fri, 6 Dec 2024 15:35:37 -0600
Subject: [PATCH] kernel32: Don't open reg keys with zero access mask.

---
 dlls/kernel32/profile.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernel32/profile.c b/dlls/kernel32/profile.c
index f65790ecc92..032802a42e0 100644
--- a/dlls/kernel32/profile.c
+++ b/dlls/kernel32/profile.c
@@ -1032,7 +1032,7 @@ static HKEY open_file_mapping_key( const WCHAR *filename )
 
     if (!mapping_key && RegOpenKeyExW( HKEY_LOCAL_MACHINE,
                                        L"Software\\Microsoft\\Windows NT\\CurrentVersion\\IniFileMapping",
-                                       0, KEY_WOW64_64KEY, &mapping_key ))
+                                       0, KEY_ENUMERATE_SUB_KEYS | KEY_WOW64_64KEY, &mapping_key ))
         mapping_key = NULL;
 
     LeaveCriticalSection( &PROFILE_CritSect );
-- 
GitLab

