From 29bcc7145aad748c5e34c9e8b56b8bf4391a3777 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Fri, 6 Dec 2024 15:50:46 -0600
Subject: [PATCH] setupapi: Don't open reg keys with zero access mask.

---
 dlls/setupapi/devinst.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/setupapi/devinst.c b/dlls/setupapi/devinst.c
index 4e406fe700e..34e9454c828 100644
--- a/dlls/setupapi/devinst.c
+++ b/dlls/setupapi/devinst.c
@@ -842,7 +842,7 @@ static void remove_device(struct device *device)
     RegDeleteKeyW(device->key, emptyW);
 
     /* delete all empty parents of the key */
-    if (!RegOpenKeyExW(HKEY_LOCAL_MACHINE, Enum, 0, 0, &enum_key))
+    if (!RegOpenKeyExW(HKEY_LOCAL_MACHINE, Enum, 0, KEY_ENUMERATE_SUB_KEYS, &enum_key))
     {
         lstrcpyW(id, device->instanceId);
 
-- 
GitLab

