From 2822d170fd8ed710a95072b6d65cf70929560c2a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 14 Sep 2024 17:24:47 +0200
Subject: [PATCH] include: Add DECLSPEC_UUID/DECLSPEC_NOVTABLE to
 MIDL_INTERFACE.

---
 include/rpcndr.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/rpcndr.h b/include/rpcndr.h
index de6a0663650..4d140492558 100644
--- a/include/rpcndr.h
+++ b/include/rpcndr.h
@@ -128,7 +128,7 @@ typedef void (__RPC_USER *NDR_NOTIFY2_ROUTINE)(boolean flag);
 # endif
 #endif
 
-#define MIDL_INTERFACE(x)   struct
+#define MIDL_INTERFACE(x)   struct DECLSPEC_UUID(x) DECLSPEC_NOVTABLE
 
 struct _MIDL_STUB_MESSAGE;
 struct _MIDL_STUB_DESC;
-- 
GitLab

