diff --git a/dlls/ntdll/unix/virtual.c b/dlls/ntdll/unix/virtual.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/virtual.c
+++ b/dlls/ntdll/unix/virtual.c
@@ -352,6 +352,7 @@ static void kernel_writewatch_reset( void *start, SIZE_T len )
 
 static void kernel_writewatch_register_range( struct file_view *view, void *base, size_t size )
 {
+#ifdef UFFDIO_WRITEPROTECT
     struct uffdio_register uffdio_register;
     struct uffdio_writeprotect wp;
 
@@ -387,6 +388,10 @@ static void kernel_writewatch_register_range( struct file_view *view, void *base
         perror("ioctl(UFFDIO_WRITEPROTECT)");
         exit(-1);
     }
+#else
+    WARN( "Wine was compiled on a Linux kernel without the UFFDIO_WRITEPROTECT feature.\n" );
+    return;
+#endif
 }
 
 static NTSTATUS kernel_soft_dirty_get_write_watches( void *base, SIZE_T size, void **addresses, ULONG_PTR *count, BOOL reset )
