From 38ac31ebdd5dbe6bca376721d4b1e7ca1526ccf1 Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Mon, 27 Oct 2025 19:19:24 +0100
Subject: [PATCH] kernelbase: Reimplement WaitForSingleObject[Ex] on top of
 NtWaitForSingleObject.

---
 dlls/kernelbase/sync.c | 15 +++++++++++++--
 1 file changed, 13 insertions(+), 2 deletions(-)

diff --git a/dlls/kernelbase/sync.c b/dlls/kernelbase/sync.c
index 31620a1aad2..3948f7ea4bd 100644
--- a/dlls/kernelbase/sync.c
+++ b/dlls/kernelbase/sync.c
@@ -395,7 +395,7 @@ BOOL WINAPI DECLSPEC_HOTPATCH UnregisterWaitEx( HANDLE handle, HANDLE event )
  */
 DWORD WINAPI DECLSPEC_HOTPATCH WaitForSingleObject( HANDLE handle, DWORD timeout )
 {
-    return WaitForMultipleObjectsEx( 1, &handle, FALSE, timeout, FALSE );
+    return WaitForSingleObjectEx( handle, timeout, FALSE );
 }
 
 
@@ -404,7 +404,18 @@ DWORD WINAPI DECLSPEC_HOTPATCH WaitForSingleObject( HANDLE handle, DWORD timeout
  */
 DWORD WINAPI DECLSPEC_HOTPATCH WaitForSingleObjectEx( HANDLE handle, DWORD timeout, BOOL alertable )
 {
-    return WaitForMultipleObjectsEx( 1, &handle, FALSE, timeout, alertable );
+    NTSTATUS status;
+    LARGE_INTEGER time;
+
+    status = NtWaitForSingleObject( normalize_std_handle( handle ), alertable,
+                                    get_nt_timeout( &time, timeout ) );
+
+    if (NT_ERROR(status))
+    {
+        SetLastError( RtlNtStatusToDosError(status) );
+        status = WAIT_FAILED;
+    }
+    return status;
 }
 
 
-- 
GitLab

