From 7061fbdf9a60ed4a9502f181acdafe88e3947659 Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Thu, 30 Oct 2025 13:06:48 +0100
Subject: [PATCH] include: Add more NT pseudo-handle definitions.

---
 include/winbase.h  | 36 +++++++++++++++++++++---------------
 include/winternl.h |  8 ++++++--
 2 files changed, 27 insertions(+), 17 deletions(-)

diff --git a/include/winbase.h b/include/winbase.h
index c96f58a61b3..4ae800bfc46 100644
--- a/include/winbase.h
+++ b/include/winbase.h
@@ -2657,6 +2657,9 @@ extern WCHAR * CDECL wine_get_dos_file_name( const char * ) __WINE_DEALLOC(HeapF
 
 #define GetCurrentProcess()   NtCurrentProcess()
 #define GetCurrentThread()    NtCurrentThread()
+#define GetCurrentProcessToken() NtCurrentProcessToken()
+#define GetCurrentThreadToken()  NtCurrentThreadToken()
+#define GetCurrentThreadEffectiveToken() NtCurrentEffectiveToken()
 #define GetCurrentProcessId() HandleToULong(NtCurrentTeb()->ClientId.UniqueProcess)
 #define GetCurrentThreadId()  HandleToULong(NtCurrentTeb()->ClientId.UniqueThread)
 
@@ -2682,6 +2685,21 @@ static FORCEINLINE DWORD WINAPI GetCurrentThreadId(void)
     return HandleToULong( ((HANDLE *)NtCurrentTeb())[9] );
 }
 
+static FORCEINLINE HANDLE WINAPI GetCurrentProcessToken(void)
+{
+    return (HANDLE)~(ULONG_PTR)3;
+}
+
+static FORCEINLINE HANDLE WINAPI GetCurrentThreadToken(void)
+{
+    return (HANDLE)~(ULONG_PTR)4;
+}
+
+static FORCEINLINE HANDLE WINAPI GetCurrentThreadEffectiveToken(void)
+{
+    return (HANDLE)~(ULONG_PTR)5;
+}
+
 static FORCEINLINE DWORD WINAPI GetLastError(void)
 {
     return *(DWORD *)((void **)NtCurrentTeb() + 13);
@@ -2703,27 +2721,15 @@ WINBASEAPI HANDLE      WINAPI GetCurrentProcess(void);
 WINBASEAPI DWORD       WINAPI GetCurrentProcessId(void);
 WINBASEAPI HANDLE      WINAPI GetCurrentThread(void);
 WINBASEAPI DWORD       WINAPI GetCurrentThreadId(void);
+WINBASEAPI HANDLE      WINAPI GetCurrentProcessToken(void);
+WINBASEAPI HANDLE      WINAPI GetCurrentThreadToken(void);
+WINBASEAPI HANDLE      WINAPI GetCurrentThreadEffectiveToken(void);
 WINBASEAPI DWORD       WINAPI GetLastError(void);
 WINBASEAPI HANDLE      WINAPI GetProcessHeap(void);
 WINBASEAPI VOID        WINAPI SetLastError(DWORD);
 
 #endif  /* __WINESRC__ */
 
-static FORCEINLINE HANDLE WINAPI GetCurrentProcessToken(void)
-{
-    return (HANDLE)~(ULONG_PTR)3;
-}
-
-static FORCEINLINE HANDLE WINAPI GetCurrentThreadToken(void)
-{
-    return (HANDLE)~(ULONG_PTR)4;
-}
-
-static FORCEINLINE HANDLE WINAPI GetCurrentThreadEffectiveToken(void)
-{
-    return (HANDLE)~(ULONG_PTR)5;
-}
-
 /* WinMain(entry point) must be declared in winbase.h. */
 /* If this is not declared, we cannot compile many sources written with C++. */
 int WINAPI WinMain(HINSTANCE,HINSTANCE,LPSTR,int);
diff --git a/include/winternl.h b/include/winternl.h
index d94b9911c74..7dea51d2f93 100644
--- a/include/winternl.h
+++ b/include/winternl.h
@@ -5284,8 +5284,12 @@ NTSYSAPI NTSTATUS  WINAPI RtlLargeIntegerToChar(const ULONGLONG *,ULONG,ULONG,PC
  * Inline functions
  */
 
-#define NtCurrentProcess() ((HANDLE)~(ULONG_PTR)0)
-#define NtCurrentThread()  ((HANDLE)~(ULONG_PTR)1)
+#define NtCurrentProcess()          ((HANDLE)~(ULONG_PTR)0)
+#define NtCurrentThread()           ((HANDLE)~(ULONG_PTR)1)
+#define NtCurrentSession()          ((HANDLE)~(ULONG_PTR)2)
+#define NtCurrentProcessToken()     ((HANDLE)~(ULONG_PTR)3)
+#define NtCurrentThreadToken()      ((HANDLE)~(ULONG_PTR)4)
+#define NtCurrentEffectiveToken()   ((HANDLE)~(ULONG_PTR)5)
 
 #define RtlFillMemory(Destination,Length,Fill) memset((Destination),(Fill),(Length))
 #define RtlMoveMemory(Destination,Source,Length) memmove((Destination),(Source),(Length))
-- 
GitLab

