From a408ab6ecb6309589423e58e63efa4fd28a0da6a Mon Sep 17 00:00:00 2001
From: Makarenko Oleg <oleg@makarenk.ooo>
Date: Tue, 20 May 2025 03:54:36 +0300
Subject: [PATCH] winebus: Fix detection of devices with no axis.

---
 dlls/winebus.sys/bus_sdl.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/dlls/winebus.sys/bus_sdl.c b/dlls/winebus.sys/bus_sdl.c
index eee95e213c0..5cec049a845 100644
--- a/dlls/winebus.sys/bus_sdl.c
+++ b/dlls/winebus.sys/bus_sdl.c
@@ -999,7 +999,8 @@ static void sdl_add_device(unsigned int index)
         desc.is_gamepad = (axis_count == 6  && button_count >= 14);
     }
 
-    for (axis_offset = 0; axis_offset < axis_count; axis_offset += (options.split_controllers ? 6 : axis_count))
+    axis_offset = 0;
+    do
     {
         NTSTATUS status;
 
@@ -1026,7 +1027,9 @@ static void sdl_add_device(unsigned int index)
         }
 
         bus_event_queue_device_created(&event_queue, &impl->unix_device, &desc);
+        axis_offset += (options.split_controllers ? 6 : axis_count);
     }
+    while (axis_offset < axis_count);
 }
 
 static void process_device_event(SDL_Event *event)
-- 
GitLab

