From 07461415f4399934a3c03c8dd6e146167bd67dfd Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Thu, 25 Mar 2021 17:53:37 +0300
Subject: [PATCH 07/16] wine.inf: Create package repository for VCLibs.140.

---
 loader/wine.inf.in | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 4df88d1f386..1506a73573b 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -91,10 +91,11 @@ AddReg=\
     MCI,\
     Misc,\
     OLE,\
+    Packages,\
     Printing,\
     Services, \
     SessionMgr,\
     Tapi,\
     ThemeManager,\
     VersionInfo,\
     LicenseInformation
@@ -117,10 +118,11 @@  AddReg=\
     MCI,\
     Misc,\
     OLE,\
+    Packages.ntamd64,\
     Printing,\
     Services, \
     SessionMgr,\
     Tapi,\
     ThemeManager,\
     VersionInfo.ntamd64,\
     LicenseInformation
@@ -142,10 +144,11 @@  AddReg=\
     MCI,\
     Misc,\
     OLE,\
+    Packages.ntarm64,\
     Printing,\
     Services, \
     SessionMgr,\
     Tapi,\
     ThemeManager,\
     VersionInfo.ntamd64,\
     LicenseInformation
@@ -163,6 +166,7 @@ AddReg=\
     MCI,\
     Misc,\
+    Packages.wow64,\
     Tapi,\
     VersionInfo,\
     LicenseInformation
 
@@ -247,6 +251,7 @@ CurrentVersion="Software\Microsoft\Windows\CurrentVersion"
 CurrentVersionNT="Software\Microsoft\Windows NT\CurrentVersion"
 FontSubStr="Software\Microsoft\Windows NT\CurrentVersion\FontSubstitutes"
 Control="System\CurrentControlSet\Control"
+Packages="Software\Classes\Local Settings\Software\Microsoft\Windows\CurrentVersion\AppModel\PackageRepository\Packages"
 
 [Classes]
 HKCR,.chm,,2,"chm.file"
@@ -597,6 +602,18 @@ [OLE]
 HKLM,"Software\Microsoft\OLE","EnableDCOM",,"Y"
 HKLM,"Software\Microsoft\OLE","EnableRemoteConnect",,"N"
 
+[Packages]
+HKLM,"%Packages%\Microsoft.VCLibs.140.00_14.0.29231.0_x86__8wekyb3d8bbwe","Path",0x00020002,"%SystemRoot%\system32"
+
+[Packages.ntamd64]
+HKLM,"%Packages%\Microsoft.VCLibs.140.00_14.0.29231.0_x64__8wekyb3d8bbwe","Path",0x00020002,"%SystemRoot%\system32"
+
+[Packages.wow64]
+HKLM,"%Packages%\Microsoft.VCLibs.140.00_14.0.29231.0_x86__8wekyb3d8bbwe","Path",0x00020002,"%SystemRoot%\syswow64"
+
+[Packages.arm64]
+HKLM,"%Packages%\Microsoft.VCLibs.140.00_14.0.29231.0_arm64__8wekyb3d8bbwe","Path",0x00020002,"%SystemRoot%\system32"
+
 [Printing]
 HKLM,%Control%\Print\Monitors\Local Port,"Driver",2,"localspl.dll"
 HKLM,%Control%\Print\Printers,"DefaultSpoolDirectory",2,"%11%\spool\printers"
-- 
2.30.2

