From 4ed2b8ed284cce9194a2a93318f7226dee29451a Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Fri, 7 Mar 2025 19:20:09 +0100
Subject: [PATCH] kernelbase: Use ProcessPriorityClass info class in
 GetPriorityClass.

---
 dlls/kernelbase/process.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/kernelbase/process.c b/dlls/kernelbase/process.c
index df3447a9014..a47e2dcc01a 100644
--- a/dlls/kernelbase/process.c
+++ b/dlls/kernelbase/process.c
@@ -821,13 +821,13 @@ BOOL WINAPI DECLSPEC_HOTPATCH GetHandleInformation( HANDLE handle, DWORD *flags
  */
 DWORD WINAPI DECLSPEC_HOTPATCH GetPriorityClass( HANDLE process )
 {
-    PROCESS_BASIC_INFORMATION pbi;
+    PROCESS_PRIORITY_CLASS priority;
 
-    if (!set_ntstatus( NtQueryInformationProcess( process, ProcessBasicInformation,
-                                                  &pbi, sizeof(pbi), NULL )))
+    if (!set_ntstatus( NtQueryInformationProcess( process, ProcessPriorityClass,
+                                                  &priority, sizeof(priority), NULL )))
         return 0;
 
-    switch (pbi.BasePriority)
+    switch (priority.PriorityClass)
     {
     case PROCESS_PRIOCLASS_IDLE: return IDLE_PRIORITY_CLASS;
     case PROCESS_PRIOCLASS_BELOW_NORMAL: return BELOW_NORMAL_PRIORITY_CLASS;
-- 
GitLab

