From 82614100833b13e729f42c39b0c6ca08d9751c65 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 30 May 2024 13:55:13 +0200
Subject: [PATCH 8/9] winewayland: Remove now unnecessary window surface
 BITMAPINFO.

---
 dlls/winewayland.drv/window_surface.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/dlls/winewayland.drv/window_surface.c b/dlls/winewayland.drv/window_surface.c
index 7cff84a5bf7..084f987df6c 100644
--- a/dlls/winewayland.drv/window_surface.c
+++ b/dlls/winewayland.drv/window_surface.c
@@ -45,7 +45,6 @@ struct wayland_window_surface
     struct window_surface header;
     struct wayland_surface *wayland_surface;
     struct wayland_buffer_queue *wayland_buffer_queue;
-    BITMAPINFO info;
 };
 
 static struct wayland_window_surface *wayland_window_surface_cast(
@@ -460,7 +459,6 @@ static struct window_surface *wayland_window_surface_create(HWND hwnd, const REC
     wws = calloc(1, sizeof(*wws));
     if (!wws) return NULL;
     if (!window_surface_init(&wws->header, &wayland_window_surface_funcs, hwnd, rect, info, 0)) goto failed;
-    wws->info = *info;
 
     TRACE("created %p hwnd %p %s\n", wws, hwnd, wine_dbgstr_rect(rect));
 
-- 
2.45.2

