From ac098761c93ec5a102de9a4d025f5bd4917acf97 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Fri, 17 Apr 2026 22:19:40 +0200
Subject: [PATCH 011/121] makedep: Quiet bison warnings.

---
 tools/makedep.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/tools/makedep.c b/tools/makedep.c
index 9b3eb0623d1..f409e1eb849 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -3095,13 +3095,13 @@ static void output_source_y( struct makefile *make, struct incl_file *source, co
     if (find_include_file( make, header ))
     {
         output( "%s: %s\n", obj_dir_path( make, header ), source->filename );
-        output( "\t%s%s -o %s.tab.$$$$.c --defines=$@ %s && rm -f %s.tab.$$$$.c\n",
+        output( "\t%s%s -Wnone -o %s.tab.$$$$.c --defines=$@ %s && rm -f %s.tab.$$$$.c\n",
                 cmd_prefix( "BISON" ), bison, obj_dir_path( make, obj ),
                 source->filename, obj_dir_path( make, obj ));
         strarray_add( &make->clean_files, header );
     }
     output( "%s.tab.c: %s\n", obj_dir_path( make, obj ), source->filename );
-    output( "\t%s%s -o $@ %s\n", cmd_prefix( "BISON" ), bison, source->filename );
+    output( "\t%s%s -Wnone -o $@ %s\n", cmd_prefix( "BISON" ), bison, source->filename );
 }
 
 
-- 
2.54.0

