From c9faf051d42a762e6c4b714df518eb36e202214b Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sun, 29 Dec 2024 23:23:41 -0800
Subject: [PATCH 09/14] configure, ntdll: Use gettid() if available.

---
 configure.ac             | 1 +
 dlls/ntdll/unix/server.c | 2 ++
 2 files changed, 3 insertions(+)

diff --git a/configure.ac b/configure.ac
index 8c183f01682..d43e5abb9e1 100644
--- a/configure.ac
+++ b/configure.ac
@@ -2114,6 +2114,7 @@ AC_CHECK_FUNCS(\
 	getauxval \
 	getifaddrs \
 	getrandom \
+ 	gettid \
 	kqueue \
 	memfd_create \
 	pipe2 \
diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index 686835d4a98..aba0a612faa 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -1572,6 +1572,8 @@ static int get_unix_tid(void)
     int ret = -1;
 #ifdef HAVE_PTHREAD_GETTHREADID_NP
     ret = pthread_getthreadid_np();
+#elif defined(HAVE_GETTID)
+    ret = gettid();
 #elif defined(linux)
     ret = syscall( __NR_gettid );
 #elif defined(__sun)
-- 
2.54.0

