From 5473de4f4925ad2f3f9331641e4f619cd86848a5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 17 Sep 2024 16:01:35 +0200
Subject: [PATCH] joy.cpl: Reduce the height of some main panel controls.

---
 dlls/joy.cpl/joy.rc | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/dlls/joy.cpl/joy.rc b/dlls/joy.cpl/joy.rc
index 2abae6d4272..24ba8162532 100644
--- a/dlls/joy.cpl/joy.rc
+++ b/dlls/joy.cpl/joy.rc
@@ -38,14 +38,14 @@ FONT 8, "Ms Shell Dlg"
 {
     PUSHBUTTON      "&Disable", IDC_BUTTON_DI_DISABLE, 200, 20, 60, 15
     PUSHBUTTON      "&Reset", IDC_BUTTON_DI_RESET, 200, 40, 60, 15
-    PUSHBUTTON      "&Enable", IDC_BUTTON_ENABLE, 200, 190, 60, 15
-    PUSHBUTTON      "&Override", IDC_BUTTON_XI_OVERRIDE, 200, 100, 60, 15
+    PUSHBUTTON      "&Override", IDC_BUTTON_XI_OVERRIDE, 200, 90, 60, 15
+    PUSHBUTTON      "&Enable", IDC_BUTTON_ENABLE, 200, 160, 60, 15
     LTEXT           "Connected (DirectInput devices)", IDC_STATIC, 10, 10, 180, 10
-    LISTBOX         IDC_DI_ENABLED_LIST, 10, 20, 180, 70, WS_TABSTOP | WS_VSCROLL | LBS_NOTIFY
-    LTEXT           "Connected (XInput devices)", IDC_STATIC, 10, 90, 180, 10
-    LISTBOX         IDC_XI_ENABLED_LIST, 10, 100, 180, 70, WS_TABSTOP | WS_VSCROLL | LBS_NOTIFY
-    LTEXT           "Disabled", IDC_STATIC, 10, 180, 180, 10
-    LISTBOX         IDC_DISABLED_LIST, 10, 190, 180, 70, WS_TABSTOP | WS_VSCROLL | LBS_NOTIFY
+    LISTBOX         IDC_DI_ENABLED_LIST, 10, 20, 180, 60, WS_TABSTOP | WS_VSCROLL | LBS_NOTIFY
+    LTEXT           "Connected (XInput devices)", IDC_STATIC, 10, 80, 180, 10
+    LISTBOX         IDC_XI_ENABLED_LIST, 10, 90, 180, 60, WS_TABSTOP | WS_VSCROLL | LBS_NOTIFY
+    LTEXT           "Disabled", IDC_STATIC, 10, 150, 180, 10
+    LISTBOX         IDC_DISABLED_LIST, 10, 160, 180, 60, WS_TABSTOP | WS_VSCROLL | LBS_NOTIFY
 }
 
 IDD_TEST_DI DIALOG 0, 0, 320, 300
-- 
GitLab

