From d0f380e2429cdd529d4621d0c22d29bfeaa924ca Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Wed, 18 Jun 2025 18:01:08 -0600
Subject: [PATCH] mmdevapi/tests: Remove a test which crashes on up to date
 Windows 11.

---
 dlls/mmdevapi/tests/render.c | 4 ----
 1 file changed, 4 deletions(-)

diff --git a/dlls/mmdevapi/tests/render.c b/dlls/mmdevapi/tests/render.c
index 236a4ff71c8..2cbbe496ba6 100644
--- a/dlls/mmdevapi/tests/render.c
+++ b/dlls/mmdevapi/tests/render.c
@@ -182,9 +182,6 @@ static void test_audioclient(void)
 
     handle = CreateEventW(NULL, FALSE, FALSE, NULL);
 
-    hr = IAudioClient_QueryInterface(ac, &IID_IUnknown, NULL);
-    ok(hr == E_POINTER, "QueryInterface(NULL) returned %08lx\n", hr);
-
     unk = (void*)(LONG_PTR)0x12345678;
     hr = IAudioClient_QueryInterface(ac, &IID_NULL, (void**)&unk);
     ok(hr == E_NOINTERFACE, "QueryInterface(IID_NULL) returned %08lx\n", hr);
@@ -197,7 +194,6 @@ static void test_audioclient(void)
         ref = IUnknown_Release(unk);
         ok(ref == 1, "Released count is %lu\n", ref);
     }
-
     hr = IAudioClient_QueryInterface(ac, &IID_IAudioClient, (void**)&unk);
     ok(hr == S_OK, "QueryInterface(IID_IAudioClient) returned %08lx\n", hr);
     if (unk)
-- 
GitLab

