From 38bff397972ef0b6cd9a5313345dc88aa992d81b Mon Sep 17 00:00:00 2001
From: Zhiyi Zhang <zzhang@codeweavers.com>
Date: Thu, 14 Aug 2025 12:06:01 +0800
Subject: [PATCH] comctl32/listview: Don't draw a focus rect when keyboard cues
 are disabled.

---
 dlls/comctl32/listview.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/comctl32/listview.c b/dlls/comctl32/listview.c
index b2856db471c..925574d9f59 100644
--- a/dlls/comctl32/listview.c
+++ b/dlls/comctl32/listview.c
@@ -1738,7 +1738,7 @@ static inline BOOL LISTVIEW_GetItemW(const LISTVIEW_INFO *infoPtr, LPLVITEMW lpL
 /* used to handle collapse main item column case */
 static inline BOOL LISTVIEW_DrawFocusRect(const LISTVIEW_INFO *infoPtr, HDC hdc)
 {
-    return (infoPtr->rcFocus.left < infoPtr->rcFocus.right) ?
+    return (COMCTL32_keyboard_cues_enabled && infoPtr->rcFocus.left < infoPtr->rcFocus.right) ?
             DrawFocusRect(hdc, &infoPtr->rcFocus) : FALSE;
 }
 
@@ -5310,7 +5310,7 @@ enddraw:
         LISTVIEW_RefreshReportGrid(infoPtr, hdc);
 
     /* Draw marquee rectangle if appropriate */
-    if (infoPtr->bMarqueeSelect)
+    if (COMCTL32_keyboard_cues_enabled && infoPtr->bMarqueeSelect)
         DrawFocusRect(hdc, &infoPtr->marqueeDrawRect);
 
     if (cdmode & CDRF_NOTIFYPOSTPAINT)
-- 
GitLab

