From 756a96fd5707cb153ae37d354e2813aa9803f5d7 Mon Sep 17 00:00:00 2001
From: yaoyongjie <yaoyongjie@uniontech.com>
Date: Thu, 5 Dec 2024 14:17:34 +0800
Subject: [PATCH] wmp: Check the return value of
 IOleClientSite_QueryInterface().

---
 dlls/wmp/oleobj.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/wmp/oleobj.c b/dlls/wmp/oleobj.c
index e227f91f8ba..5309194f289 100644
--- a/dlls/wmp/oleobj.c
+++ b/dlls/wmp/oleobj.c
@@ -157,7 +157,7 @@ static HRESULT activate_inplace(WindowsMediaPlayer *This)
         IOleInPlaceSiteEx_OnInPlaceActivateEx(ipsiteex, &redraw, 0);
         ipsite = (IOleInPlaceSite*)ipsiteex;
     }else {
-        IOleClientSite_QueryInterface(This->client_site, &IID_IOleInPlaceSite, (void**)&ipsite);
+        hres = IOleClientSite_QueryInterface(This->client_site, &IID_IOleInPlaceSite, (void**)&ipsite);
         if(FAILED(hres)) {
             FIXME("No IOleInPlaceSite instance\n");
             return hres;
-- 
GitLab

