From 8be04266fbc0410f50277d7ed1999dab65ee9735 Mon Sep 17 00:00:00 2001
From: Eugene McArdle <eugene@tensorworks.com.au>
Date: Mon, 9 Dec 2024 09:28:26 +1000
Subject: [PATCH] ntdll: Simplify check that unexpected files were not found

---
 dlls/ntdll/tests/directory.c | 11 ++---------
 1 file changed, 2 insertions(+), 9 deletions(-)

diff --git a/dlls/ntdll/tests/directory.c b/dlls/ntdll/tests/directory.c
index b07654ee1ff..b98d399d7e2 100644
--- a/dlls/ntdll/tests/directory.c
+++ b/dlls/ntdll/tests/directory.c
@@ -1041,15 +1041,8 @@ static BOOL test_NtQueryDirectoryFile_mask(HANDLE handle, BOOL restart_scan, UNI
     else
     {
         if (validate_only && status == STATUS_SUCCESS) return FALSE;
-        /* Print the query parameters and the result */
-        if (status == 0) {
-            dir_info = (FILE_DIRECTORY_INFORMATION *)data;
-            name = dir_info->FileName;
-            name_len = dir_info->FileNameLength / sizeof(WCHAR);
-
-            ok ( status != STATUS_SUCCESS, "unexpectedly found file. mask: '%s', found %s\n",
-                wine_dbgstr_wn(mask.Buffer, mask_len), wine_dbgstr_wn(name, name_len) );
-        }
+        
+        ok ( status == STATUS_NO_MORE_FILES, "unexpectedly found file\n" );
     }
     return TRUE;
 }
-- 
GitLab

