From 08163fb73e94978a4e15605d2ae2209774253b25 Mon Sep 17 00:00:00 2001
From: Mohamad Al-Jaf <mohamadaljaf@gmail.com>
Date: Wed, 16 Apr 2025 20:21:22 -0400
Subject: [PATCH] windows.media.mediacontrol: Fix a memory leak.

---
 dlls/windows.media.mediacontrol/main.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/windows.media.mediacontrol/main.c b/dlls/windows.media.mediacontrol/main.c
index 18d7d3ce5f9..e8c74ed6ed3 100644
--- a/dlls/windows.media.mediacontrol/main.c
+++ b/dlls/windows.media.mediacontrol/main.c
@@ -178,6 +178,7 @@ static ULONG WINAPI music_properties_Release( IMusicDisplayProperties *iface )
 
     if (!ref)
     {
+        WindowsDeleteString( impl->album_title );
         WindowsDeleteString( impl->artist );
         WindowsDeleteString( impl->title );
         free( impl );
-- 
GitLab

