From 001137a9c547013303b4feeec73df61f31ce4c4b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?D=C4=81vis=20Mos=C4=81ns?= <davispuh@gmail.com>
Date: Fri, 1 Dec 2023 04:44:30 +0200
Subject: [PATCH] kernelbase: LoadLibraryEx validate flags like Win10

---
 dlls/kernelbase/loader.c | 12 ++++++++++++
 include/winbase.h        |  4 ++++
 2 files changed, 16 insertions(+)

diff --git a/dlls/kernelbase/loader.c b/dlls/kernelbase/loader.c
index af61428487b..316b1d180cb 100644
--- a/dlls/kernelbase/loader.c
+++ b/dlls/kernelbase/loader.c
@@ -537,6 +537,18 @@ HMODULE WINAPI DECLSPEC_HOTPATCH LoadLibraryExW( LPCWSTR name, HANDLE file, DWOR
                         ((flags & LOAD_LIBRARY_AS_DATAFILE) && (flags & LOAD_LIBRARY_AS_DATAFILE_EXCLUSIVE)) ||
                         ((flags & LOAD_LIBRARY_SEARCH_DEFAULT_DIRS) && (flags & LOAD_WITH_ALTERED_SEARCH_PATH));
 
+    /* Windows 10 22H2 have such validation */
+    invalid_parameter = invalid_parameter ||
+                        ((flags & LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR) &&
+                            (flags & (LOAD_LIBRARY_AS_DATAFILE | LOAD_LIBRARY_AS_DATAFILE_EXCLUSIVE | LOAD_LIBRARY_AS_IMAGE_RESOURCE)));
+
+    if (flags & LOAD_LIBRARY_UNSUPPORTED_MASK)
+    {
+        WARN( "Pretending to be Windows 10 22H2, unsupported flags: 0x%lx\n, returning ERROR_INVALID_PARAMETER",
+              flags & LOAD_LIBRARY_UNSUPPORTED_MASK );
+        invalid_parameter = TRUE;
+    }
+
     if (invalid_parameter)
     {
         SetLastError( ERROR_INVALID_PARAMETER );
diff --git a/include/winbase.h b/include/winbase.h
index 93b592c3791..bcfd0fc8d1f 100644
--- a/include/winbase.h
+++ b/include/winbase.h
@@ -1006,6 +1006,10 @@ DECL_WINELIB_TYPE_AW(ENUMRESLANGPROC)
 #define LOAD_LIBRARY_SEARCH_USER_DIRS       0x00000400
 #define LOAD_LIBRARY_SEARCH_SYSTEM32        0x00000800
 #define LOAD_LIBRARY_SEARCH_DEFAULT_DIRS    0x00001000
+#define LOAD_LIBRARY_SAFE_CURRENT_DIRS      0x00002000
+#define LOAD_LIBRARY_SEARCH_SYSTEM32_NO_FORWARDER 0x00004000
+#define LOAD_LIBRARY_OS_INTEGRITY_CONTINUITY 0x00008000
+#define LOAD_LIBRARY_UNSUPPORTED_MASK       0xFFFF0000 /* If you add new flags update this mask */
 
 #define GET_MODULE_HANDLE_EX_FLAG_PIN                 1
 #define GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT  2
-- 
GitLab

