From 4751074016f0b8c514d3d218b56da5679e3fea44 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Thu, 13 Jun 2019 12:58:27 -0500
Subject: [PATCH 0526/1533] server: Add a diagnostic message for fsync.

---
 server/fsync.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/server/fsync.c b/server/fsync.c
index 6a67d22d14e..d4dbd66d22a 100644
--- a/server/fsync.c
+++ b/server/fsync.c
@@ -124,6 +124,8 @@ void fsync_init(void)
 
     is_fsync_initialized = 1;
 
+    fprintf( stderr, "fsync: up and running.\n" );
+
     atexit( shm_cleanup );
 }
 
-- 
2.45.0

