From 3ed94a67cf5626372e4732b5d58faa8245c15916 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Wed, 18 Mar 2020 20:03:42 -0500
Subject: [PATCH 0540/1533] ntdll: Default the spin count to 100.

---
 dlls/ntdll/unix/fsync.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/unix/fsync.c b/dlls/ntdll/unix/fsync.c
index da5245f9e85..c28b4e27597 100644
--- a/dlls/ntdll/unix/fsync.c
+++ b/dlls/ntdll/unix/fsync.c
@@ -89,7 +89,7 @@ static inline int futex_wait( int *addr, int val, struct timespec *timeout )
     return syscall( __NR_futex, addr, 0, val, timeout, 0, 0 );
 }
 
-static unsigned int spincount;
+static unsigned int spincount = 100;
 
 int do_fsync(void)
 {
-- 
2.45.0

