From d86b58e1f5a54a267dabfbfc75dafb93a1678b40 Mon Sep 17 00:00:00 2001
From: Derek Lesho <dlesho@codeweavers.com>
Date: Mon, 14 Feb 2022 12:51:27 -0500
Subject: [PATCH 0551/1533] fsync: Type-check HANDLE in fsync_set_event().

Signed-off-by: Derek Lesho <dlesho@codeweavers.com>
---
 dlls/ntdll/unix/fsync.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/ntdll/unix/fsync.c b/dlls/ntdll/unix/fsync.c
index 193bd37e6b8..f9b75fd2faf 100644
--- a/dlls/ntdll/unix/fsync.c
+++ b/dlls/ntdll/unix/fsync.c
@@ -576,6 +576,9 @@ NTSTATUS fsync_set_event( HANDLE handle, LONG *prev )
     if ((ret = get_object( handle, &obj ))) return ret;
     event = obj->shm;
 
+    if (obj->type != FSYNC_MANUAL_EVENT && obj->type != FSYNC_AUTO_EVENT)
+        return STATUS_OBJECT_TYPE_MISMATCH;
+
     if (!(current = __atomic_exchange_n( &event->signaled, 1, __ATOMIC_SEQ_CST )))
         futex_wake( &event->signaled, INT_MAX );
 
-- 
2.45.0

