diff --git a/programs/wineboot/wineboot.c b/programs/wineboot/wineboot.c
index a934748394c..62c1f5f3fe5 100644
--- a/programs/wineboot/wineboot.c
+++ b/programs/wineboot/wineboot.c
@@ -914,7 +914,7 @@ static void create_bios_system_values( HKEY bios_key, const char *buf, UINT len
     }
 }
 
-static void create_bios_processor_values( HKEY system_key, const char *buf, UINT len )
+static void create_bios_processor_values( HKEY system_key, const char *buf, UINT len, UINT64 tsc_frequency )
 {
     const struct smbios_header *hdr;
     const struct smbios_processor *proc;
@@ -923,7 +923,6 @@ static void create_bios_processor_values( HKEY system_key, const char *buf, UINT
     SYSTEM_CPU_INFORMATION sci;
     PROCESSOR_POWER_INFORMATION* power_info;
     ULONG sizeof_power_info = sizeof(PROCESSOR_POWER_INFORMATION) * NtCurrentTeb()->Peb->NumberOfProcessors;
-    UINT64 tsc_frequency = read_tsc_frequency();
     const WCHAR *arch;
     WCHAR id[60], buffer[128], *version, *vendorid;
 
@@ -1069,7 +1068,7 @@ static void create_hardware_registry_keys( UINT64 tsc_frequency )
         create_bios_system_values( bios_key, buf, len );
         RegCloseKey( bios_key );
     }
-    create_bios_processor_values( system_key, buf, len );
+    create_bios_processor_values( system_key, buf, len, tsc_frequency );
     RegCloseKey( system_key );
     free( buf );
 }
