From c2e936a008399c7a97086bea1a3e0ff6836a5c15 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sun, 6 Apr 2025 09:16:48 -0700
Subject: [PATCH] ntdll: Avoid a syscall for RtlGetSystemTimePrecise.

---
 dlls/ntdll/time.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/time.c b/dlls/ntdll/time.c
index 7ac45d20deb..5420aea25d7 100644
--- a/dlls/ntdll/time.c
+++ b/dlls/ntdll/time.c
@@ -373,9 +373,15 @@ void WINAPI RtlTimeToElapsedTimeFields( const LARGE_INTEGER *Time, PTIME_FIELDS
 LONGLONG WINAPI RtlGetSystemTimePrecise( void )
 {
     LONGLONG ret;
+    ULONG high, low;
+    do
+    {
+        high = user_shared_data->SystemTime.High1Time;
+        low = user_shared_data->SystemTime.LowPart;
+    }
+    while (high != user_shared_data->SystemTime.High2Time);
 
-    WINE_UNIX_CALL( unix_system_time_precise, &ret );
-    return ret;
+    return (ULONGLONG)high << 32 | low;
 }
 
 /* 128-bit multiply a by b and return the high 64 bits, same as __umulh */
-- 
2.49.0

