From d3024005ea36c1c81206582ccab74cd4c342f2f6 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 20 May 2025 11:35:50 +0200
Subject: [PATCH] server: Pass NULL as child region if region is NULL.

---
 server/window.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/server/window.c b/server/window.c
index c59b8646118..9d59e0837bc 100644
--- a/server/window.c
+++ b/server/window.c
@@ -1690,7 +1690,7 @@ static void redraw_window( struct window *win, struct region *region, unsigned i
             if (rect_in_region( child_rgn, &child->window_rect ))
             {
                 offset_region( child_rgn, -child->client_rect.left, -child->client_rect.top );
-                redraw_window( child, child_rgn, flags, 1 );
+                redraw_window( child, region ? child_rgn : NULL, flags, 1 );
             }
         }
         free_region( child_rgn );
-- 
GitLab

