From 61ebc6c4c203e95c8f0f2242a9287659d5111742 Mon Sep 17 00:00:00 2001
From: winebuild <wine@build.dev>
Date: Sat, 22 Feb 2025 00:10:58 -0800
Subject: [PATCH] ntdll: crash hotfix

---
 dlls/ntdll/loader.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index a5cf6a49467..20acdd14793 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -2576,7 +2576,7 @@ static ULONG read_image_directory( HANDLE file, const SECTION_IMAGE_INFORMATION
     offset.QuadPart += offsetof( IMAGE_NT_HEADERS32, OptionalHeader ) + nt.nt32.FileHeader.SizeOfOptionalHeader;
     count = nt.nt32.FileHeader.NumberOfSections;
     if (!(sec = RtlAllocateHeap( GetProcessHeap(), 0, sizeof(*sec) * count ))) return 0;
-    if (NtReadFile( file, 0, NULL, NULL, &io, &sec, count * sizeof(*sec), &offset, NULL )) goto done;
+    if (NtReadFile( file, 0, NULL, NULL, &io, sec, count * sizeof(*sec), &offset, NULL )) goto done;
     if (io.Information != count * sizeof(*sec)) goto done;
     for (i = 0; i < count; i++)
     {
-- 
2.48.1

