From b63bcb68f61910196b58d7e7d2d0f59ea9884a26 Mon Sep 17 00:00:00 2001
From: Haoyang Chen <chenhaoyang@kylinos.cn>
Date: Tue, 6 Aug 2024 15:35:01 +0800
Subject: [PATCH] wined3d: Fix a memory leak.

---
 dlls/wined3d/context_gl.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/wined3d/context_gl.c b/dlls/wined3d/context_gl.c
index 75f94678e14..cc7f766b20a 100644
--- a/dlls/wined3d/context_gl.c
+++ b/dlls/wined3d/context_gl.c
@@ -1286,6 +1286,7 @@ static BOOL wined3d_context_gl_set_gl_context(struct wined3d_context_gl *context
             return FALSE;
         }
 
+        wined3d_release_dc(context_gl->window, context_gl->dc);
         if (!(context_gl->dc = wined3d_device_gl_get_backup_dc(device_gl)))
         {
             wined3d_context_gl_set_current(NULL);
-- 
GitLab

