From 8120f0591cbaeedd451cc2674a5465945945aa13 Mon Sep 17 00:00:00 2001
From: SmithGoll <2022209065@qq.com>
Date: Thu, 15 Aug 2024 23:04:47 +0000
Subject: [PATCH] explorer: Fix desktop command-line parameter parsing.

---
 programs/explorer/desktop.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/programs/explorer/desktop.c b/programs/explorer/desktop.c
index fb59258ad36..fb521188717 100644
--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -1158,11 +1158,19 @@ void manage_desktop( WCHAR *arg )
     }
 
     /* parse the desktop option */
-    /* the option is of the form /desktop=name[,widthxheight[,driver]] */
+    /* the option is of the form /desktop[=name[,widthxheight[,driver]]] */
     if ((arg[0] == '=' || arg[0] == ',') && arg[1] && arg[1] != ',')
     {
-        arg++;
-        name = arg;
+        if (arg[0] == ',')
+        {
+            name = get_default_desktop_name();
+        }
+        else
+        {
+            arg++;
+            name = arg;
+        }
+
         if ((p = wcschr( arg, ',' )))
         {
             *p++ = 0;
-- 
GitLab

