From 34e9931b28d75d98aa35d302c9ce0abcfc3c98d8 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sun, 6 Apr 2025 09:01:37 -0700
Subject: [PATCH] kernelbase: Cache locale in SetThreadLocale.

---
 dlls/kernelbase/thread.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/dlls/kernelbase/thread.c b/dlls/kernelbase/thread.c
index eecafd164a1..fe736078ebc 100644
--- a/dlls/kernelbase/thread.c
+++ b/dlls/kernelbase/thread.c
@@ -529,11 +529,16 @@ BOOL WINAPI DECLSPEC_HOTPATCH SetThreadIdealProcessorEx( HANDLE thread, PROCESSO
  */
 BOOL WINAPI DECLSPEC_HOTPATCH SetThreadLocale( LCID lcid )
 {
-    lcid = ConvertDefaultLocale( lcid );
+    static LCID cached_lcid = -1, cached_unconverted_lcid = -1;
+    if (cached_lcid == NtCurrentTeb()->CurrentLocale && cached_unconverted_lcid == lcid)
+        return TRUE;
+    cached_unconverted_lcid = lcid;
+    lcid = cached_lcid = ConvertDefaultLocale( lcid );
     if (lcid != GetThreadLocale())
     {
         if (!IsValidLocale( lcid, LCID_SUPPORTED ))
         {
+            cached_lcid = -1;
             SetLastError( ERROR_INVALID_PARAMETER );
             return FALSE;
         }
-- 
2.49.0

