From b0515fbb61f7118ed24db134106604d63d618a11 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sun, 11 Dec 2022 20:09:44 +0100
Subject: [PATCH 66/68] ntdll: Avoid writing to invalid memory in i386 unix
 dispatcher.


 100.0% dlls/ntdll/unix/
diff --git a/dlls/ntdll/unix/signal_i386.c b/dlls/ntdll/unix/signal_i386.c
index 751b0081534..d86857be6e3 100644
--- a/dlls/ntdll/unix/signal_i386.c
+++ b/dlls/ntdll/unix/signal_i386.c
@@ -2779,8 +2779,7 @@ __ASM_GLOBAL_FUNC( __wine_unix_call_dispatcher,
                    __ASM_CFI_REG_IS_AT1(esi, ecx, 0x30)
                    "movl %ebp,0x34(%ecx)\n\t"
                    __ASM_CFI_REG_IS_AT1(ebp, ecx, 0x34)
-                   "movl 12(%esp),%edx\n\t"    /* args */
-                   "movl %edx,-16(%ecx)\n\t"
+                   "movl 12(%esp),%ebx\n\t"    /* args */
                    "movl (%esp),%eax\n\t"      /* handle */
                    "movl 8(%esp),%edx\n\t"     /* code */
                    /* switch to kernel stack */
@@ -2792,10 +2791,13 @@ __ASM_GLOBAL_FUNC( __wine_unix_call_dispatcher,
                    __ASM_CFI(".cfi_offset %ebx,-12\n\t")
                    __ASM_CFI(".cfi_offset %esi,-16\n\t")
                    __ASM_CFI(".cfi_offset %edi,-20\n\t")
+                   "movl %ebx,(%esp)\n\t"
                    "call *(%eax,%edx,4)\n\t"
                    "leal 16(%esp),%esp\n\t"
                    "testw $0xffff,2(%esp)\n\t" /* frame->restore_flags */
                    "jnz " __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") "\n\t"
+                   "movl 0x20(%esp),%ebx\n\t"
+                   __ASM_CFI(".cfi_same_value %ebx\n\t")
                    "movl 0x08(%esp),%ecx\n\t"  /* frame->eip */
                    /* switch to user stack */
                    "movl 0x0c(%esp),%esp\n\t"  /* frame->esp */
-- 
2.43.0

