From 42e22e7f5fc29942908584aaa015407c7a4e548b Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 14 May 2024 14:53:55 +0300
Subject: [PATCH] win32u: Clip PAINTSTRUCT.rcPaint to the window client area.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 dlls/win32u/dce.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/dlls/win32u/dce.c b/dlls/win32u/dce.c
index 64788990707..6be1a489c38 100644
--- a/dlls/win32u/dce.c
+++ b/dlls/win32u/dce.c
@@ -1594,6 +1594,8 @@ static BOOL send_erase( HWND hwnd, UINT flags, HRGN client_rgn,
     HDC hdc = 0;
     RECT dummy;
 
+    TRACE( "hwnd %p, flags %08x, client_rgn %p\n", hwnd, flags, client_rgn );
+
     if (!clip_rect) clip_rect = &dummy;
     if (hdc_ret || (flags & UPDATE_ERASE))
     {
@@ -1604,6 +1606,13 @@ static BOOL send_erase( HWND hwnd, UINT flags, HRGN client_rgn,
         {
             INT type = NtGdiGetAppClipBox( hdc, clip_rect );
 
+            if (get_class_long( hwnd, GCL_STYLE, FALSE ) & CS_PARENTDC)
+            {
+                RECT client_rect;
+                get_client_rect( hwnd, &client_rect, get_thread_dpi() );
+                intersect_rect( clip_rect, clip_rect, &client_rect );
+            }
+
             if (flags & UPDATE_ERASE)
             {
                 /* don't erase if the clip box is empty */
-- 
GitLab

