From cf165dfcd1b797fc57a6a793b3e2d906b1cbc61d Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Sun, 23 Aug 2026 01:48:34 +0200
Subject: [PATCH 17/17] winex11: Look up the XInputExtension opcode at startup.

---
 dlls/winex11.drv/x11drv_main.c | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/dlls/winex11.drv/x11drv_main.c b/dlls/winex11.drv/x11drv_main.c
index e38deeea249..82ba33c5386 100644
--- a/dlls/winex11.drv/x11drv_main.c
+++ b/dlls/winex11.drv/x11drv_main.c
@@ -40,6 +40,9 @@
 #ifdef HAVE_X11_EXTENSIONS_XRENDER_H
 #include <X11/extensions/Xrender.h>
 #endif
+#ifdef HAVE_X11_EXTENSIONS_XINPUT_H
+#include <X11/extensions/XInput.h>
+#endif
 
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
@@ -90,6 +93,10 @@ int xfixes_event_base = 0;
 char *process_name = NULL;
 HANDLE steam_overlay_event;
 
+#ifdef HAVE_X11_EXTENSIONS_XINPUT_H
+static int xinput_opcode = 0;   /* major opcode of XInputExtension */
+#endif
+
 static BOOL err_locked_display;
 static x11drv_error_callback err_callback;   /* current callback for error */
 static Display *err_callback_display;        /* display callback is set for */
@@ -244,8 +251,10 @@ static inline BOOL ignore_error( Display *display, XErrorEvent *event )
     }
 
 #ifdef HAVE_X11_EXTENSIONS_XINPUT_H
-    if (event->request_code == 131 /* XInputExtension */ &&
-        event->minor_code   == 3   /* X_OpenDevice */)
+    /* query the major opcode of an extension since it seems to be assigned by the server at runtime */
+    if (xinput_opcode &&
+        event->request_code == xinput_opcode &&
+        event->minor_code   == 3 /* X_OpenDevice */)
     {
         return TRUE;
     }
@@ -721,6 +730,23 @@ static void init_visuals( Display *display, int screen )
            default_visual.visualid, default_visual.class, argb_visual.visualid );
 }
 
+#ifdef HAVE_X11_EXTENSIONS_XINPUT_H
+/***********************************************************************
+ *           x11drv_xinput_init
+ *
+ * Retrieve the major opcode of XInputExtension.
+ */
+static void x11drv_xinput_init( Display *display )
+{
+    int event_base, error_base;
+
+    if (!XQueryExtension( display, "XInputExtension", &xinput_opcode, &event_base, &error_base ))
+        xinput_opcode = 0;
+
+    TRACE( "XInput extension major opcode %d\n", xinput_opcode );
+}
+#endif
+
 /***********************************************************************
  *           X11DRV process initialisation routine
  */
@@ -793,6 +819,9 @@ static NTSTATUS x11drv_init( void *arg )
     abs_tablet_hack = atoi( getenv( "WINE_ABS_TABLET" ) ?:
                             getenv( "WINE_ENABLE_ABS_TABLET_HACK" ) ?: "0" ) != 0;
     x11drv_xinput2_load();
+#ifdef HAVE_X11_EXTENSIONS_XINPUT_H
+    x11drv_xinput_init( display );
+#endif
 
     XkbUseExtension( gdi_display, NULL, NULL );
     X11DRV_InitKeyboard( gdi_display );
-- 
2.55.0

