From 2d000d05e050d33c928dfecb4ebcafa35179483d Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Tue, 31 Oct 2023 12:23:30 -0500
Subject: [PATCH 02/32] gcc 13 warnings

---
 dlls/webservices/tests/channel.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/dlls/webservices/tests/channel.c b/dlls/webservices/tests/channel.c
index 4cf39c10732..1fb12297eef 100644
--- a/dlls/webservices/tests/channel.c
+++ b/dlls/webservices/tests/channel.c
@@ -1214,6 +1214,9 @@ static const char send_record_begin[] = {
 static const char send_record_middle[] = { 0x01, 0x56, 0x0e, 0x42 };
 static const char send_record_end[] = { 0x08, 0x02, 0x6e, 0x73, 0x89, 0xff, 0x01, 0x01 };
 
+#pragma GCC diagnostic ignored "-Warray-bounds"
+#pragma GCC diagnostic ignored "-Wstringop-overflow"
+
 static BOOL send_dict_str( int sock, char *addr, const char *str, int dict_str_count )
 {
     char buf[512], dict_buf[256], body_buf[128], dict_size_buf[5];
-- 
2.47.1


