From e50ade335386be619330ff9940bd9406e97333b7 Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Sat, 13 Mar 2021 16:20:30 -0600
Subject: [PATCH 28/32] server: Allow disabling in-process synchronization
 support.

---
 server/inproc_sync.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/server/inproc_sync.c b/server/inproc_sync.c
index c50bbb52bbc..e708aefe249 100644
--- a/server/inproc_sync.c
+++ b/server/inproc_sync.c
@@ -128,6 +128,12 @@ static struct linux_device *get_linux_device(void)
     struct linux_device *device;
     int unix_fd;
 
+    if (getenv( "WINE_DISABLE_FAST_SYNC" ) && atoi( getenv( "WINE_DISABLE_FAST_SYNC" ) ))
+    {
+        set_error( STATUS_NOT_IMPLEMENTED );
+        return NULL;
+    }
+
     if (linux_device_object)
         return (struct linux_device *)grab_object( linux_device_object );
 
-- 
2.47.1


