From 1fc22e8a54318d88408a3d0eb6655311874d9fd1 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 23 Aug 2024 19:47:05 +0200
Subject: [PATCH] mfplat: Seek byte stream to the start for URL hint detection.

---
 dlls/mfplat/main.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index aa987ff0496..6c4ee89ba8b 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -6257,8 +6257,8 @@ static HRESULT resolver_get_bytestream_url_hint(IMFByteStream *stream, WCHAR con
 
     if (FAILED(hr = IMFByteStream_GetCurrentPosition(stream, &position)))
         return hr;
-
-    hr = IMFByteStream_Read(stream, buffer, sizeof(buffer), &length);
+    if (SUCCEEDED(hr = IMFByteStream_SetCurrentPosition(stream, 0)))
+        hr = IMFByteStream_Read(stream, buffer, sizeof(buffer), &length);
     IMFByteStream_SetCurrentPosition(stream, position);
     if (FAILED(hr))
         return hr;
-- 
GitLab

