From 0be1f1db312f087077c6a827ce58267475d6fe49 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 21 Oct 2025 00:21:22 +0200
Subject: [PATCH 11/17] win32u: Hoist get_thread_dpi in apply_window_pos.

---
 dlls/win32u/window.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index 60663ecb87e..3bd9c3fd487 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -2110,16 +2110,16 @@ static BOOL apply_window_pos( HWND hwnd, HWND insert_after, UINT swp_flags, stru
     struct window_rects old_rects;
     RECT extra_rects[3];
     struct window_surface *old_surface;
-    UINT raw_dpi, monitor_dpi;
+    UINT raw_dpi, monitor_dpi, dpi = get_thread_dpi();
 
     is_layered = new_surface && new_surface->alpha_mask;
     is_fullscreen = is_window_rect_full_screen( &new_rects->visible, get_thread_dpi() );
     is_child = parent && parent != NtUserGetDesktopWindow();
 
     if (is_child) monitor_dpi = get_win_monitor_dpi( parent, &raw_dpi );
-    else monitor_dpi = monitor_dpi_from_rect( new_rects->window, get_thread_dpi(), &raw_dpi );
+    else monitor_dpi = monitor_dpi_from_rect( new_rects->window, dpi, &raw_dpi );
 
-    get_window_rects( hwnd, COORDS_PARENT, &old_rects, get_thread_dpi() );
+    get_window_rects( hwnd, COORDS_PARENT, &old_rects, dpi );
     if (IsRectEmpty( &valid_rects[0] ) || is_layered) valid_rects = NULL;
 
     if (!(win = get_win_ptr( hwnd )) || win == WND_DESKTOP || win == WND_OTHER_PROCESS) return FALSE;
@@ -2133,8 +2133,8 @@ static BOOL apply_window_pos( HWND hwnd, HWND insert_after, UINT swp_flags, stru
         valid_rects = NULL;
     }
 
-    if (is_child) monitor_rects = map_dpi_window_rects( *new_rects, get_thread_dpi(), raw_dpi );
-    else monitor_rects = map_window_rects_virt_to_raw( *new_rects, get_thread_dpi() );
+    if (is_child) monitor_rects = map_dpi_window_rects( *new_rects, dpi, raw_dpi );
+    else monitor_rects = map_window_rects_virt_to_raw( *new_rects, dpi );
 
     SERVER_START_REQ( set_window_pos )
     {
@@ -2171,7 +2171,7 @@ static BOOL apply_window_pos( HWND hwnd, HWND insert_after, UINT swp_flags, stru
             if (get_window_long( win->parent, GWL_EXSTYLE ) & WS_EX_LAYOUTRTL)
             {
                 RECT client = {0};
-                get_client_rect_rel( win->parent, COORDS_CLIENT, &client, get_thread_dpi() );
+                get_client_rect_rel( win->parent, COORDS_CLIENT, &client, dpi );
                 mirror_rect( &client, &win->rects.window );
                 mirror_rect( &client, &win->rects.client );
                 mirror_rect( &client, &win->rects.visible );
-- 
2.51.2

