From 1ad7904ee0c8c14d93e30430f96736aff610fa87 Mon Sep 17 00:00:00 2001
From: Mohamad Al-Jaf <mohamadaljaf@gmail.com>
Date: Wed, 12 Mar 2025 21:57:39 -0400
Subject: [PATCH] windows.system.profile.systemid/tests: Add
 ISystemIdentificationStatics::GetSystemIdForPublisher() tests.

---
 .../tests/systemid.c                              | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/dlls/windows.system.profile.systemid/tests/systemid.c b/dlls/windows.system.profile.systemid/tests/systemid.c
index 1187b6f438f..44a2682f593 100644
--- a/dlls/windows.system.profile.systemid/tests/systemid.c
+++ b/dlls/windows.system.profile.systemid/tests/systemid.c
@@ -47,6 +47,7 @@ static void test_SystemIdentification_Statics(void)
 {
     static const WCHAR *system_id_statics_name = L"Windows.System.Profile.SystemIdentification";
     ISystemIdentificationStatics *system_id_statics = (void *)0xdeadbeef;
+    ISystemIdentificationInfo *system_id_info = (void *)0xdeadbeef;
     IActivationFactory *factory = (void *)0xdeadbeef;
     HSTRING str = NULL;
     HRESULT hr;
@@ -70,6 +71,20 @@ static void test_SystemIdentification_Statics(void)
     hr = IActivationFactory_QueryInterface( factory, &IID_ISystemIdentificationStatics, (void **)&system_id_statics );
     ok( hr == S_OK, "got hr %#lx.\n", hr );
 
+    hr = ISystemIdentificationStatics_GetSystemIdForPublisher( system_id_statics, NULL );
+    todo_wine
+    ok( hr == E_INVALIDARG, "got hr %#lx.\n", hr );
+    hr = ISystemIdentificationStatics_GetSystemIdForPublisher( system_id_statics, &system_id_info );
+    todo_wine
+    ok( hr == S_OK, "got hr %#lx.\n", hr );
+    if (SUCCEEDED(hr))
+    {
+        check_interface( system_id_info, &IID_IAgileObject, FALSE );
+
+        ref = ISystemIdentificationInfo_Release( system_id_info );
+        ok( ref == 0, "got ref %ld.\n", ref );
+    }
+
     ref = ISystemIdentificationStatics_Release( system_id_statics );
     ok( ref == 2, "got ref %ld.\n", ref );
     ref = IActivationFactory_Release( factory );
-- 
GitLab

