From 169689879e09c0666f2d9edf55ef2054084bd1b8 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Mon, 16 Dec 2024 05:22:34 -0800
Subject: [PATCH] win32u: Use fast user_shared_data to check for driver updates
 at 10khz.

---
 dlls/win32u/message.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/dlls/win32u/message.c b/dlls/win32u/message.c
index 59ea115d260..48771f74bdf 100644
--- a/dlls/win32u/message.c
+++ b/dlls/win32u/message.c
@@ -26,6 +26,8 @@
 #endif
 
 #include <assert.h>
+#include "winternl.h"
+#include "ddk/wdm.h"
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "win32u_private.h"
@@ -40,6 +41,8 @@ WINE_DEFAULT_DEBUG_CHANNEL(msg);
 WINE_DECLARE_DEBUG_CHANNEL(key);
 WINE_DECLARE_DEBUG_CHANNEL(relay);
 
+static const struct _KUSER_SHARED_DATA *user_shared_data = (struct _KUSER_SHARED_DATA *)0x7ffe0000;
+
 #define MAX_WINPROC_RECURSION  64
 
 #define WM_NCMOUSEFIRST WM_NCMOUSEMOVE
@@ -3064,9 +3068,7 @@ static HANDLE get_server_queue_handle(void)
 /* monotonic timer tick for throttling driver event checks */
 static inline LONGLONG get_driver_check_time(void)
 {
-    LARGE_INTEGER counter, freq;
-    NtQueryPerformanceCounter( &counter, &freq );
-    return counter.QuadPart * 8000 / freq.QuadPart; /* 8kHz */
+    return user_shared_data->InterruptTime.LowPart / 1000;
 }
 
 /* check for driver events if we detect that the app is not properly consuming messages */
-- 
2.47.1

