diff --git a/dlls/ntdll/unix/sync.c b/dlls/ntdll/unix/sync.c
index 91bab77489a..8bae3cfe686 100644
--- a/dlls/ntdll/unix/sync.c
+++ b/dlls/ntdll/unix/sync.c
@@ -1848,17 +1848,7 @@ NTSTATUS WINAPI NtSignalAndWaitForSingleObject( HANDLE signal, HANDLE wait,
 NTSTATUS WINAPI NtYieldExecution(void)
 {
 #ifdef HAVE_SCHED_YIELD
-#ifdef RUSAGE_THREAD
-    struct rusage u1, u2;
-    int ret;
-
-    ret = getrusage( RUSAGE_THREAD, &u1 );
-#endif
     sched_yield();
-#ifdef RUSAGE_THREAD
-    if (!ret) ret = getrusage( RUSAGE_THREAD, &u2 );
-    if (!ret && u1.ru_nvcsw == u2.ru_nvcsw && u1.ru_nivcsw == u2.ru_nivcsw) return STATUS_NO_YIELD_PERFORMED;
-#endif
     return STATUS_SUCCESS;
 #else
     return STATUS_NO_YIELD_PERFORMED;
