From cb50cc2638e671174ea91828440600a50d744652 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 20 Feb 2025 04:09:12 -0800
Subject: [PATCH] server: Use separate time variables for the USD thread.

---
 server/fd.c | 14 +++++++-------
 1 file changed, 7 insertions(+), 7 deletions(-)

diff --git a/server/fd.c b/server/fd.c
index 3834b5ec456..3d75ea83dad 100644
--- a/server/fd.c
+++ b/server/fd.c
@@ -1053,10 +1053,10 @@ static int get_next_timeout(void)
 static void *update_user_shared_data_time_thread(void *param)
 {
     struct timespec ts;
-    timeout_t next_update_time;
+    timeout_t next_update_time, usd_current_time, usd_monotonic_time;
 
-    get_current_time(&current_time, &monotonic_time);
-    next_update_time = monotonic_time + user_shared_data_timeout_ticks;
+    get_current_time(&usd_current_time, &usd_monotonic_time);
+    next_update_time = usd_monotonic_time + user_shared_data_timeout_ticks;
 
     while (active_users)
     {
@@ -1064,13 +1064,13 @@ static void *update_user_shared_data_time_thread(void *param)
 
         while (clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &ts, NULL) == EINTR);
 
-        get_current_time(&current_time, &monotonic_time);
-        set_user_shared_data_time(current_time, monotonic_time);
+        get_current_time(&usd_current_time, &usd_monotonic_time);
+        set_user_shared_data_time(usd_current_time, usd_monotonic_time);
 
         next_update_time += user_shared_data_timeout_ticks;
 
-        if (next_update_time <= monotonic_time)
-            next_update_time = monotonic_time + user_shared_data_timeout_ticks;
+        if (next_update_time <= usd_monotonic_time)
+            next_update_time = usd_monotonic_time + user_shared_data_timeout_ticks;
     }
     return NULL;
 }
-- 
2.48.1

