From ef1de504fd7a71338e0fb4c0ecad52f890490496 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 27 Feb 2025 09:26:03 -0800
Subject: [PATCH 058/121] winepulse: Align stream buffer to 64 bytes.

---
 dlls/winepulse.drv/pulse.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 7f9cc93ad9b..ca438140f7b 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -1209,12 +1209,13 @@ static NTSTATUS pulse_create_stream(void *args)
         return STATUS_SUCCESS;
     }
 
-    if (!(stream = calloc(1, sizeof(*stream))))
+    if (!(stream = aligned_alloc(64, sizeof(*stream))))
     {
         pulse_unlock();
         params->result = E_OUTOFMEMORY;
         return STATUS_SUCCESS;
     }
+    memset(stream, 0, sizeof(*stream));
 
     stream->dataflow = params->flow;
     for (i = 0; i < ARRAY_SIZE(stream->vol); ++i)
-- 
2.54.0

