From 8c2842d8970a03a7aa938b70f96c9a8c2e74b95b Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 27 Feb 2025 09:26:03 -0800
Subject: [PATCH 06/10] winepulse: Align stream buffer to 64 bytes.

---
 dlls/winepulse.drv/pulse.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 551a7247904..debb5697513 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -1191,12 +1191,13 @@ static NTSTATUS pulse_create_stream(void *args)
         return STATUS_SUCCESS;
     }
 
-    if (!(stream = calloc(1, sizeof(*stream))))
+    if (!(stream = aligned_alloc(64, sizeof(*stream))))
     {
         pulse_unlock();
         params->result = E_OUTOFMEMORY;
         return STATUS_SUCCESS;
     }
+    memset(stream, 0, sizeof(*stream));
 
     stream->dataflow = params->flow;
     for (i = 0; i < ARRAY_SIZE(stream->vol); ++i)
-- 
2.54.0

