From 462658f795c856b3b3cb9f77c092cb1e306c0fc2 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sun, 16 Mar 2025 17:31:41 -0700
Subject: [PATCH 059/121] dsound: Divide default buffer length by a factor of
 8.

---
 dlls/dsound/dsound_main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/dsound/dsound_main.c b/dlls/dsound/dsound_main.c
index e2b9b81beef..0b5c3f42b81 100644
--- a/dlls/dsound/dsound_main.c
+++ b/dlls/dsound/dsound_main.c
@@ -70,7 +70,7 @@ GUID *DSOUND_renderer_guids;
 GUID *DSOUND_capture_guids;
 
 /* All default settings, you most likely don't want to touch these, see wiki on UsefulRegistryKeys */
-int ds_hel_buflen = 32768 * 2;
+int ds_hel_buflen = 4096 * 2;
 
 /*
  * Get a config key from either the app-specific or the default config
-- 
2.54.0

