From a9c4d54624d7c00b2c562071c38cce18c5e6722b Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Tue, 10 Sep 2024 11:21:17 -0700
Subject: [PATCH] combase: Check null token earlier.

---
 dlls/combase/combase.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/dlls/combase/combase.c b/dlls/combase/combase.c
index d478b30beb0..b363bc771ef 100644
--- a/dlls/combase/combase.c
+++ b/dlls/combase/combase.c
@@ -2655,6 +2655,8 @@ HRESULT WINAPI CoGetContextToken(ULONG_PTR *token)
 
     TRACE("%p\n", token);
 
+    if(!token) return E_POINTER;
+
     if (!InternalIsProcessInitialized())
     {
         ERR("apartment not initialised\n");
@@ -2670,9 +2672,6 @@ HRESULT WINAPI CoGetContextToken(ULONG_PTR *token)
     if (FAILED(hr = com_get_tlsdata(&tlsdata)))
         return hr;
 
-    if (!token)
-        return E_POINTER;
-
     if (!tlsdata->context_token)
     {
         struct thread_context *context;
-- 
2.46.0

