From 2881bc2bcc6b22fd0e109f4e97050e20e1bf17bc Mon Sep 17 00:00:00 2001
From: Joshua Ashton <joshua@froggi.es>
Date: Thu, 16 Jul 2020 10:18:31 +0100
Subject: [PATCH 24/47] user32: Draw buttons as flat by default

Signed-off-by: Joshua Ashton <joshua@froggi.es>
---
 dlls/user32/uitools.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/dlls/user32/uitools.c b/dlls/user32/uitools.c
index a8f3a0af84b..5f3d9dd51b7 100644
--- a/dlls/user32/uitools.c
+++ b/dlls/user32/uitools.c
@@ -167,6 +167,8 @@ static BOOL UITOOLS95_DrawDiagEdge(HDC hdc, LPRECT rc,
     int add = (LTRBInnerMono[uType & (BDR_INNER|BDR_OUTER)] != -1 ? 1 : 0)
             + (LTRBOuterMono[uType & (BDR_INNER|BDR_OUTER)] != -1 ? 1 : 0);
 
+    uFlags |= BF_FLAT;
+
     /* Init some vars */
     OuterPen = InnerPen = GetStockObject(NULL_PEN);
     SavePen = SelectObject(hdc, InnerPen);
@@ -474,6 +476,8 @@ static BOOL UITOOLS95_DrawRectEdge(HDC hdc, LPRECT rc,
                        || (uType & BDR_OUTER) == BDR_OUTER)
                       && !(uFlags & (BF_FLAT|BF_MONO)) );
 
+    uFlags |= BF_FLAT;
+
     lti_brush = lto_brush = rbi_brush = rbo_brush = GetStockObject(NULL_BRUSH);
 
     /* Determine the colors of the edges */
@@ -1492,7 +1496,7 @@ BOOL WINAPI DrawFocusRect( HDC hdc, const RECT* rc )
     hOldBrush = SelectObject(hdc, GetStockObject(NULL_BRUSH));
     lb.lbStyle = BS_SOLID;
     lb.lbColor = 0;
-    hNewPen = ExtCreatePen(PS_COSMETIC|PS_ALTERNATE, 1, &lb, 0, NULL);
+    hNewPen = ExtCreatePen(PS_COSMETIC, 1, &lb, 0, NULL);
     hOldPen = SelectObject(hdc, hNewPen);
     oldDrawMode = SetROP2(hdc, R2_NOT);
     oldBkMode = SetBkMode(hdc, TRANSPARENT);
-- 
2.54.0

