From 68dcb76964e8bac78bf829d0b392b5d8ab948087 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Sat, 18 Apr 2026 01:27:06 +0200
Subject: [PATCH 108/121] ntdll: Quieten SYSTEM_PERFORMANCE_INFORMATION.

---
 dlls/ntdll/unix/system.c | 6 +-----
 1 file changed, 1 insertion(+), 5 deletions(-)

diff --git a/dlls/ntdll/unix/system.c b/dlls/ntdll/unix/system.c
index 2c033a4b15c..2762c314a17 100644
--- a/dlls/ntdll/unix/system.c
+++ b/dlls/ntdll/unix/system.c
@@ -3296,7 +3296,6 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
     case SystemPerformanceInformation:  /* 2 */
     {
         SYSTEM_PERFORMANCE_INFORMATION spi;
-        static BOOL fixme_written = FALSE;
 
         get_performance_info( &spi );
         len = sizeof(spi);
@@ -3306,10 +3305,7 @@ NTSTATUS WINAPI NtQuerySystemInformation( SYSTEM_INFORMATION_CLASS class,
             else memcpy( info, &spi, len);
         }
         else ret = STATUS_INFO_LENGTH_MISMATCH;
-        if(!fixme_written) {
-            FIXME("info_class SYSTEM_PERFORMANCE_INFORMATION\n");
-            fixme_written = TRUE;
-        }
+        WARN("info_class SYSTEM_PERFORMANCE_INFORMATION\n");
         break;
     }
 
-- 
2.54.0

