From 1ecb70ecf584bca18d63177b358710f107a3ad3e Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Sat, 18 Apr 2026 01:33:09 +0200
Subject: [PATCH 119/121] win32u: Add winstation cache.

---
 dlls/win32u/winstation.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/dlls/win32u/winstation.c b/dlls/win32u/winstation.c
index 3da9acdd3af..e9c01fb3904 100644
--- a/dlls/win32u/winstation.c
+++ b/dlls/win32u/winstation.c
@@ -419,19 +419,25 @@ BOOL WINAPI NtUserCloseWindowStation( HWINSTA handle )
     return ret;
 }
 
+static HWINSTA winsta_cache = INVALID_HANDLE_VALUE;
+
 /***********************************************************************
  *           NtUSerGetProcessWindowStation  (win32u.@)
  */
 HWINSTA WINAPI NtUserGetProcessWindowStation(void)
 {
+    HWINSTA cache = winsta_cache;
     HWINSTA ret = 0;
 
+    if (cache != INVALID_HANDLE_VALUE) return cache;
+
     SERVER_START_REQ( get_process_winstation )
     {
         if (!wine_server_call_err( req ))
             ret = wine_server_ptr_handle( reply->handle );
     }
     SERVER_END_REQ;
+    winsta_cache = ret;
     return ret;
 }
 
@@ -442,6 +448,8 @@ BOOL WINAPI NtUserSetProcessWindowStation( HWINSTA handle )
 {
     BOOL ret;
 
+    winsta_cache = INVALID_HANDLE_VALUE;
+
     SERVER_START_REQ( set_process_winstation )
     {
         req->handle = wine_server_obj_handle( handle );
-- 
2.54.0

