diff --git a/dlls/win32u/rawinput.c b/dlls/win32u/rawinput.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/rawinput.c
+++ b/dlls/win32u/rawinput.c
@@ -327,9 +327,25 @@ static struct device *find_device_from_handle( HANDLE handle, BOOL refresh )
  */
 UINT WINAPI NtUserGetRawInputDeviceList( RAWINPUTDEVICELIST *device_list, UINT *device_count, UINT size )
 {
+    static UINT cache, count_saved;
+    static RAWINPUTDEVICELIST *device_list_saved;
     unsigned int count = 0;
     struct device *device;
 
+    if (cache) {
+        if (device_list == device_list_saved) {
+            return count_saved;
+        }
+        else if (!device_list) {
+            device_list = device_list_saved;
+            return count_saved;
+        }
+        else {
+            // invalidate cache
+            cache = count_saved = 0;
+        }
+    }
+
     TRACE( "device_list %p, device_count %p, size %u.\n", device_list, device_count, size );
 
     if (size != sizeof(*device_list))
@@ -358,6 +374,12 @@ UINT WINAPI NtUserGetRawInputDeviceList( RAWINPUTDEVICELIST *device_list, UINT *
 
     pthread_mutex_unlock( &rawinput_mutex );
 
+    if (device_list && !cache) {
+        cache++;
+        device_list_saved = device_list;
+        count_saved = count;
+    }
+
     if (!device_list)
     {
         *device_count = count;
