diff --git a/dlls/win32u/imm.c b/dlls/win32u/imm.c
index 11111111111..11111111111 100644
--- a/dlls/win32u/imm.c
+++ b/dlls/win32u/imm.c
@@ -281,7 +281,8 @@ BOOL register_imm_window( HWND hwnd )
 
     TRACE( "(%p)\n", hwnd );
 
-    if (disable_ime || !needs_ime_window( hwnd ))
+    const char *e = getenv("WINE_DISABLE_IME");
+    if ((e && *e != '\0' && *e != '0') || disable_ime || !needs_ime_window( hwnd ))
         return FALSE;
 
     thread_data = get_imm_thread_data();
