diff --git a/dlls/winex11.drv/xinerama.c b/dlls/winex11.drv/xinerama.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/xinerama.c
+++ b/dlls/winex11.drv/xinerama.c
@@ -179,7 +179,7 @@ BOOL xinerama_get_fullscreen_monitors( const RECT *rect, long *indices )
     }
 
     if (indices[0] == -1 || indices[1] == -1 || indices[2] == -1 || indices[3] == -1)
-        ERR("Failed to get xinerama fullscreen monitor indices.\n");
+        WARN("Failed to get xinerama fullscreen monitor indices.\n");
     else
         ret = TRUE;
 
