diff --git a/dlls/winex11.drv/xinerama.c b/dlls/winex11.drv/xinerama.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/xinerama.c
+++ b/dlls/winex11.drv/xinerama.c
@@ -179,7 +179,7 @@ BOOL xinerama_get_fullscreen_monitors( const RECT *rect, long *indices )
     }
 
     if (indices[0] == -1 || indices[1] == -1 || indices[2] == -1 || indices[3] == -1)
-        ERR("Failed to get xinerama fullscreen monitor indices.\n");
+        WARN("Failed to get xinerama fullscreen monitor indices.\n");
     else
         ret = TRUE;
 
diff --git a/dlls/ntdll/unix/env.c b/dlls/ntdll/unix/env.c
index 11111111111..11111111111 100644
--- a/dlls/ntdll/unix/env.c
+++ b/dlls/ntdll/unix/env.c
@@ -1884,7 +1884,7 @@ static void init_peb( RTL_USER_PROCESS_PARAMETERS *params, void *module )
         NtCurrentTeb()->Tib.ExceptionList = (void *)((char *)NtCurrentTeb() + teb_offset);
         wow_peb = (PEB32 *)((char *)peb + page_size);
         set_thread_id( NtCurrentTeb(), GetCurrentProcessId(), GetCurrentThreadId() );
-        ERR( "starting %s in experimental wow64 mode\n", debugstr_us(&params->ImagePathName) );
+        WARN( "starting %s in experimental wow64 mode\n", debugstr_us(&params->ImagePathName) );
         break;
     case IMAGE_FILE_MACHINE_AMD64:
         if (main_image_info.Machine == current_machine) break;
