diff --git a/dlls/winex11.drv/window.c b/dlls/winex11.drv/window.c
index 11111111111..11111111111 100644
--- a/dlls/winex11.drv/window.c
+++ b/dlls/winex11.drv/window.c
@@ -2902,7 +2902,7 @@ BOOL X11DRV_CreateWindow( HWND hwnd )
         static const WCHAR winsta0[] = {'W','i','n','S','t','a','0',0};
         struct x11drv_thread_data *data = x11drv_init_thread_data();
         WCHAR winstation_name[64];
-        XSetWindowAttributes attr;
+        XSetWindowAttributes attr = {0};

         if (NtUserGetObjectInformation( NtUserGetProcessWindowStation(), UOI_NAME, winstation_name,
                                         sizeof(winstation_name), NULL ))
