diff --git a/dlls/winewayland.drv/wayland_output.c b/dlls/winewayland.drv/wayland_output.c
index f76881a1770..0a89a5be199 100644
--- a/dlls/winewayland.drv/wayland_output.c
+++ b/dlls/winewayland.drv/wayland_output.c
@@ -71,6 +71,20 @@ static void wayland_output_state_add_mode(struct wayland_output_state *state,
                                           int32_t width, int32_t height,
                                           int32_t refresh, BOOL current)
 {
+    static BOOL wayland_disable_60hz;
+
+    if (!wayland_disable_60hz && refresh < 70000) {
+        const char *e = getenv("WINE_WAYLAND_NO60HZ");
+        if (e && *e != '\0' && *e != '0') {
+            wayland_disable_60hz = TRUE;
+
+        }
+    }
+    if (wayland_disable_60hz && refresh < 70000) { 
+        FIXME("Found 60hz mode, not adding it\n");
+        return; 
+    }
+
     struct rb_entry *mode_entry;
     struct wayland_output_mode *mode;
     struct wayland_output_mode key =
