From 1e2ddbef1b586269afbc4611f8009dcbe308a16d Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 14 May 2024 14:42:00 +0300
Subject: [PATCH] server: If the being validated region covers whole window
 then validate everything.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 server/window.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/server/window.c b/server/window.c
index f7f9d5e517f..d36d54e3c9c 100644
--- a/server/window.c
+++ b/server/window.c
@@ -1641,7 +1641,24 @@ static void redraw_window( struct window *win, struct region *region, int frame,
         {
             if ((tmp = crop_region_to_win_rect( win, region, frame )))
             {
-                if (!subtract_region( tmp, win->update_region, tmp ))
+                if ((child_rgn = create_empty_region()))
+                {
+                    struct rectangle rect = win->window_rect;
+
+                    offset_rect( &rect, -rect.left, -rect.top );
+                    set_region_rect( child_rgn, &rect );
+
+                    if (subtract_region( child_rgn, child_rgn, tmp ) && is_region_empty( child_rgn ) )
+                    {
+                        /* region covers whole window: validate everything */
+                        free_region( tmp );
+                        tmp = NULL;
+                    }
+
+                    free_region( child_rgn );
+                }
+
+                if (tmp && !subtract_region( tmp, win->update_region, tmp ))
                 {
                     free_region( tmp );
                     return;
-- 
GitLab

