From 881511d2fcafe23eafe5f12c6adcb850d18ed2ba Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Tue, 14 May 2024 14:46:12 +0300
Subject: [PATCH] server: For a CS_PARENTDC child use parent for visible region
 calculations.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 server/window.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/server/window.c b/server/window.c
index d36d54e3c9c..b42f4de1049 100644
--- a/server/window.c
+++ b/server/window.c
@@ -1425,6 +1425,18 @@ static struct region *crop_region_to_win_rect( struct window *win, struct region
     struct region *tmp;
 
     if (!get_window_visible_rect( win, &rect, frame )) return NULL;
+
+    if (win->parent && is_window_using_parent_dc( win ))
+    {
+        int offset_x, offset_y;
+
+        if (!get_window_visible_rect( win->parent, &rect, 0 )) return NULL;
+
+        offset_x = win->parent->client_rect.left - win->parent->window_rect.left + win->window_rect.left;
+        offset_y = win->parent->client_rect.top - win->parent->window_rect.top + win->window_rect.top;
+        offset_rect( &rect, -offset_x, -offset_y );
+    }
+
     if (!(tmp = create_empty_region())) return NULL;
     set_region_rect( tmp, &rect );
 
-- 
GitLab

