From bd6efa43ac180661825ca55af1b83e49efd5a591 Mon Sep 17 00:00:00 2001
From: Dmitry Timoshkov <dmitry@baikal.ru>
Date: Wed, 7 May 2025 17:36:18 +0300
Subject: [PATCH] server: Don't generate WM_PAINT messages for minimized
 windows.

Signed-off-by: Dmitry Timoshkov <dmitry@baikal.ru>
---
 server/window.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/server/window.c b/server/window.c
index b42f4de1049..04793bf035e 100644
--- a/server/window.c
+++ b/server/window.c
@@ -1236,6 +1236,7 @@ static struct region *get_visible_region( struct window *win, unsigned int flags
     /* first check if all ancestors are visible */
 
     if (!is_visible( win )) return region;  /* empty region */
+    if (win->style & WS_MINIMIZE) return region;
 
     if (is_desktop_window( win ))
     {
@@ -1401,6 +1402,7 @@ static int get_window_visible_rect( struct window *win, struct rectangle *rect,
     *rect = frame ? win->window_rect : win->client_rect;
 
     if (!(win->style & WS_VISIBLE)) return 0;
+    if (win->style & WS_MINIMIZE) return 0;
     if (is_desktop_window( win )) return 1;
 
     while (!is_desktop_window( win->parent ))
-- 
GitLab

