From 0cafdfb20f9a1350723f0200afd40d74ca85f19b Mon Sep 17 00:00:00 2001
From: Maotong Zhang <zmtong1988@gmail.com>
Date: Thu, 24 Jul 2025 10:54:31 +0800
Subject: [PATCH] oleaut32/test: VarAnd(VT_BSTR,VT_I2), VT_BSTR length exceed
 VT_I2 fail

---
 dlls/oleaut32/tests/vartest.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/dlls/oleaut32/tests/vartest.c b/dlls/oleaut32/tests/vartest.c
index eb1639e3990..6593f64201b 100644
--- a/dlls/oleaut32/tests/vartest.c
+++ b/dlls/oleaut32/tests/vartest.c
@@ -7451,6 +7451,8 @@ static void test_VarAnd(void)
     VARAND(BSTR,false_str,BSTR,false_str,BOOL,0);
     VARAND(BSTR,true_str,BSTR,false_str,BOOL,VARIANT_FALSE);
     VARAND(BSTR,true_str,BSTR,true_str,BOOL,VARIANT_TRUE);
+    VARAND(BSTR, SysAllocString(L"2147483647"), I2, 15, I4, (2147483647 & 15));
+    VARAND(BSTR, SysAllocString(L"9223372036854775807"), I2, 15, I8, (9223372036854775807 & 15));
     VARANDCY(BSTR,true_str,10000,I4,1);
     VARANDCY(BSTR,false_str,10000,I4,0);
 
-- 
GitLab

