From 87fd1100d245fb62ae654d184f0411a32addb1d8 Mon Sep 17 00:00:00 2001
From: panhui <panhui@uniontech.com>
Date: Thu, 2 Jul 2026 16:48:22 -0400
Subject: [PATCH 52/59] winewayland: Treat 0x0 windows as roleless.

---
 dlls/winewayland.drv/window.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/winewayland.drv/window.c b/dlls/winewayland.drv/window.c
index 39e69ea8c5f..7b7cd91c021 100644
--- a/dlls/winewayland.drv/window.c
+++ b/dlls/winewayland.drv/window.c
@@ -218,7 +218,7 @@ static BOOL wayland_win_data_create_wayland_surface(struct wayland_win_data *dat
     if (!(surface = data->wayland_surface) || !surface->window.visible)
         visible = visible && layer_set;
 
-    if (!visible) role = WAYLAND_SURFACE_ROLE_NONE;
+    if (!visible || IsRectEmpty(&data->rects.visible)) role = WAYLAND_SURFACE_ROLE_NONE;
     else if (owner_surface) role = WAYLAND_SURFACE_ROLE_POPUP;
     else role = WAYLAND_SURFACE_ROLE_TOPLEVEL;
 
