From 25d6c724e1a7b1456a582285c17dbee612282cfb Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Thu, 30 Oct 2025 21:42:58 +0100
Subject: [PATCH 39/44] kernelbase: Use NT_ERROR() to check for errors in
 WaitForMultipleObjectsEx.

---
 dlls/kernelbase/sync.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernelbase/sync.c b/dlls/kernelbase/sync.c
index 6c69a0ab88f..72f3a1796c4 100644
--- a/dlls/kernelbase/sync.c
+++ b/dlls/kernelbase/sync.c
@@ -438,7 +438,7 @@ DWORD WINAPI DECLSPEC_HOTPATCH WaitForMultipleObjectsEx( DWORD count, const HAND
 
     status = NtWaitForMultipleObjects( count, hloc, !wait_all, alertable,
                                        get_nt_timeout( &time, timeout ) );
-    if (HIWORD(status))  /* is it an error code? */
+    if (NT_ERROR(status))
     {
         SetLastError( RtlNtStatusToDosError(status) );
         status = WAIT_FAILED;
-- 
2.51.2

