From cfb84f5092225c1ee3623dc2a95436f0c90529e8 Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Mon, 14 Jul 2025 16:17:09 -0600
Subject: [PATCH] ntdll: Use NtContinueEx in KiUserApcDispatcher on x64.

---
 dlls/ntdll/signal_x86_64.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
index 31070566d40..1f184f4ea58 100644
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -377,8 +377,8 @@ __ASM_GLOBAL_FUNC( KiUserApcDispatcher,
                    "movq %rsp,%r9\n\t"         /* context */
                    "callq *%rax\n\t"
                    "movq %rsp,%rcx\n\t"        /* context */
-                   "movl $1,%edx\n\t"          /* alertable */
-                   "call " __ASM_NAME("NtContinue") "\n\t"
+                   "leaq 0x4f0(%rsp),%rdx\n\t" /* continue_arg */
+                   "call " __ASM_NAME("NtContinueEx") "\n\t"
                    "int3" )
 
 
-- 
GitLab

