From 5fa6c52ce78f6f4b73fcb757d6f4f5db70cfd973 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Tue, 4 Nov 2025 19:34:37 +0100
Subject: [PATCH] opengl32: Simplify extension tokenization.

---
 dlls/opengl32/unix_wgl.c | 35 ++++++++++++++---------------------
 1 file changed, 14 insertions(+), 21 deletions(-)

diff --git a/dlls/opengl32/unix_wgl.c b/dlls/opengl32/unix_wgl.c
index 4f346b39226..ea3119195b8 100644
--- a/dlls/opengl32/unix_wgl.c
+++ b/dlls/opengl32/unix_wgl.c
@@ -1233,29 +1233,22 @@ static void make_context_current( TEB *teb, const struct opengl_funcs *funcs, HD
     }
     else
     {
-        const char *str = (const char *)funcs->p_glGetString( GL_EXTENSIONS );
-        size_t len = strlen( str );
-        const char *p;
-        char *ext;
-        if (!str) str = "";
-        if ((len = strlen( str )) && str[len - 1] == ' ') len--;
-        if (*str) size++;
-        for (p = str; p < str + len; p++) if (*p == ' ') size++;
-        if (!(extensions = malloc( size * sizeof(*extensions) + len + 1 ))) return;
-        ext = (char *)&extensions[size];
-        memcpy( ext, str, len );
-        ext[len] = 0;
-        if (*ext) extensions[count++] = ext;
-        while (*ext)
+        const char *str, *ptr;
+        char *out;
+
+        if (!(str = (const char *)funcs->p_glGetString( GL_EXTENSIONS ))) str = "";
+        for (ptr = str; *ptr; ptr++) if (ptr == str || ptr[-1] == ' ') size++;
+
+        if (!(extensions = malloc( size * sizeof(*extensions) + (ptr - str + 1) ))) return;
+
+        for (out = (char *)&extensions[size], ptr = str; *ptr; out++, ptr++)
         {
-            if (*ext == ' ')
-            {
-                *ext = 0;
-                extensions[count++] = ext + 1;
-            }
-            ext++;
+            if (ptr == str || ptr[-1] == ' ') extensions[count++] = out;
+            *out = *ptr == ' ' ? 0 : *ptr;
         }
-        assert( count + ARRAYSIZE(legacy_extensions) - 1 == size );
+        if (ptr != str) *out = 0;
+
+        assert( count + ARRAY_SIZE(legacy_extensions) - 1 == size );
     }
 
     if (!disabled && !(disabled = query_opengl_option( "DisabledExtensions" ))) disabled = "";
-- 
GitLab

