From 70af8770cd425bd8be5f2ddadde6774bfd3cb3e8 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Wed, 31 Jul 2019 12:18:37 -0500
Subject: [PATCH 0534/2346] server: Print a message when using server-side
 synchronization.

---
 server/main.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/server/main.c b/server/main.c
index 457b044a36a..9559ad83584 100644
--- a/server/main.c
+++ b/server/main.c
@@ -237,6 +237,9 @@ int main( int argc, char *argv[] )
     if (do_esync())
         esync_init();
 
+    if (!do_fsync() && !do_esync())
+        fprintf( stderr, "wineserver: using server-side synchronization.\n" );
+
     if (debug_level) fprintf( stderr, "wineserver: starting (pid=%ld)\n", (long) getpid() );
     set_current_time();
     init_signals();
-- 
2.47.0

