From 8d0f0f800176d4b28d885de5034bd7b55303c19b Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Thu, 16 Jan 2025 11:00:53 -0600
Subject: [PATCH] crypt32: Factor out CRYPT_RegDeleteFromReg().

---
 dlls/crypt32/regstore.c | 21 ++++++++++-----------
 1 file changed, 10 insertions(+), 11 deletions(-)

diff --git a/dlls/crypt32/regstore.c b/dlls/crypt32/regstore.c
index 8567604c39b..962bd6a0382 100644
--- a/dlls/crypt32/regstore.c
+++ b/dlls/crypt32/regstore.c
@@ -237,6 +237,15 @@ BOOL CRYPT_SerializeContextsToReg(HKEY key, DWORD flags,
     return ret;
 }
 
+static void CRYPT_RegDeleteFromReg(HKEY key, const BYTE *sha1_hash)
+{
+    WCHAR hash[20 * 2 + 1];
+
+    CRYPT_HashToStr(sha1_hash, hash);
+    TRACE("Removing %s\n", debugstr_w(hash));
+    RegDeleteKeyW(key, hash);
+}
+
 static BOOL CRYPT_RegWriteToReg(WINE_REGSTOREINFO *store)
 {
     static const WCHAR * const subKeys[] = { L"Certificates", L"CRLs", L"CTLs" };
@@ -258,22 +267,12 @@ static BOOL CRYPT_RegWriteToReg(WINE_REGSTOREINFO *store)
             if (listToDelete[i])
             {
                 WINE_HASH_TO_DELETE *toDelete, *next;
-                WCHAR asciiHash[20 * 2 + 1];
 
                 EnterCriticalSection(&store->cs);
                 LIST_FOR_EACH_ENTRY_SAFE(toDelete, next, listToDelete[i],
                  WINE_HASH_TO_DELETE, entry)
                 {
-                    LONG rc;
-
-                    CRYPT_HashToStr(toDelete->hash, asciiHash);
-                    TRACE("Removing %s\n", debugstr_w(asciiHash));
-                    rc = RegDeleteKeyW(key, asciiHash);
-                    if (rc != ERROR_SUCCESS && rc != ERROR_FILE_NOT_FOUND)
-                    {
-                        SetLastError(rc);
-                        ret = FALSE;
-                    }
+                    CRYPT_RegDeleteFromReg(key, toDelete->hash);
                     list_remove(&toDelete->entry);
                     CryptMemFree(toDelete);
                 }
-- 
GitLab

