From a6ce5c75b2ca0beffee03606e90d5a7f56602403 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 1 Jul 2024 11:03:57 +0200
Subject: [PATCH 4/8] winewayland: Rely on win32u previous surface reuse.

---
 dlls/winewayland.drv/window_surface.c | 18 +++---------------
 1 file changed, 3 insertions(+), 15 deletions(-)

diff --git a/dlls/winewayland.drv/window_surface.c b/dlls/winewayland.drv/window_surface.c
index e9eb74b17ca..ecbc023833e 100644
--- a/dlls/winewayland.drv/window_surface.c
+++ b/dlls/winewayland.drv/window_surface.c
@@ -512,29 +512,17 @@ void wayland_window_surface_update_wayland_surface(struct window_surface *window
  */
 BOOL WAYLAND_CreateWindowSurface(HWND hwnd, const RECT *surface_rect, struct window_surface **surface)
 {
+    struct window_surface *previous;
     struct wayland_win_data *data;
 
     TRACE("hwnd %p, surface_rect %s, surface %p\n", hwnd, wine_dbgstr_rect(surface_rect), surface);
 
+    if ((previous = *surface) && previous->funcs == &wayland_window_surface_funcs) return TRUE;
     if (!(data = wayland_win_data_get(hwnd))) return TRUE; /* use default surface */
-
-    /* Release the dummy surface wine provides for toplevels. */
-    if (*surface) window_surface_release(*surface);
-    *surface = NULL;
-
-    /* Check if we can reuse our current window surface. */
-    if (data->window_surface &&
-        EqualRect(&data->window_surface->rect, surface_rect))
-    {
-        window_surface_add_ref(data->window_surface);
-        *surface = data->window_surface;
-        TRACE("reusing surface %p\n", *surface);
-        goto done;
-    }
+    if (previous) window_surface_release(previous);
 
     *surface = wayland_window_surface_create(data->hwnd, surface_rect);
 
-done:
     wayland_win_data_release(data);
     return TRUE;
 }
-- 
2.45.2

