From f0006b6133fe7544091c58d989ae967effa065a2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?D=C4=81vis=20Mos=C4=81ns?= <davispuh@gmail.com>
Date: Fri, 1 Dec 2023 04:36:25 +0200
Subject: [PATCH] kernelbase: LoadLibraryEx support LOAD_PACKAGED_LIBRARY

---
 dlls/kernelbase/loader.c | 7 +++++++
 include/winbase.h        | 1 +
 2 files changed, 8 insertions(+)

diff --git a/dlls/kernelbase/loader.c b/dlls/kernelbase/loader.c
index 8f40286b880..af61428487b 100644
--- a/dlls/kernelbase/loader.c
+++ b/dlls/kernelbase/loader.c
@@ -543,6 +543,13 @@ HMODULE WINAPI DECLSPEC_HOTPATCH LoadLibraryExW( LPCWSTR name, HANDLE file, DWOR
         return 0;
     }
 
+    if (flags & LOAD_PACKAGED_LIBRARY)
+    {
+        /* FIXME: Don't know if it's implemented like this on Windows
+           but this does accomplish same result in error case */
+        return LoadPackagedLibrary(name, flags);
+    }
+
     RtlInitUnicodeString( &str, name );
     if (str.Buffer[str.Length/sizeof(WCHAR) - 1] != ' ') return load_library( &str, flags );
 
diff --git a/include/winbase.h b/include/winbase.h
index 16e449f7c2d..93b592c3791 100644
--- a/include/winbase.h
+++ b/include/winbase.h
@@ -995,6 +995,7 @@ DECL_WINELIB_TYPE_AW(ENUMRESLANGPROC)
 /* flags that can be passed to LoadLibraryEx */
 #define DONT_RESOLVE_DLL_REFERENCES         0x00000001
 #define LOAD_LIBRARY_AS_DATAFILE            0x00000002
+#define LOAD_PACKAGED_LIBRARY               0x00000004 /* Undocumented Windows internal */
 #define LOAD_WITH_ALTERED_SEARCH_PATH       0x00000008
 #define LOAD_IGNORE_CODE_AUTHZ_LEVEL        0x00000010
 #define LOAD_LIBRARY_AS_IMAGE_RESOURCE      0x00000020
-- 
GitLab

