From 40f6d3bb66d355a127b3da231461b76d69960a61 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sun, 15 Jun 2025 14:15:34 +0200
Subject: [PATCH] user32: Move user32_module default out of get_class_info.

---
 dlls/user32/class.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/user32/class.c b/dlls/user32/class.c
index d0dd09b441f..9be0c131b26 100644
--- a/dlls/user32/class.c
+++ b/dlls/user32/class.c
@@ -561,8 +561,6 @@ ATOM get_class_info( HINSTANCE instance, const WCHAR *class_name, WNDCLASSEXW *i
     init_class_name( &name, class_name );
     get_class_version( &name, NULL, &module );
 
-    if (!name_str && !instance) instance = user32_module;
-
     while (!(atom = NtUserGetClassInfoEx( instance, &name, info, NULL, ansi )))
     {
         if (module)
@@ -610,6 +608,7 @@ BOOL WINAPI GetClassInfoExA( HINSTANCE hInstance, LPCSTR name, WNDCLASSEXA *wc )
         return FALSE;
     }
 
+    if (!hInstance) hInstance = user32_module;
     if (!IS_INTRESOURCE(name))
     {
         WCHAR nameW[MAX_ATOM_LEN + 1];
@@ -640,6 +639,7 @@ BOOL WINAPI GetClassInfoExW( HINSTANCE hInstance, LPCWSTR name, WNDCLASSEXW *wc
         return FALSE;
     }
 
+    if (!hInstance) hInstance = user32_module;
     atom = get_class_info( hInstance, name, wc, NULL, FALSE );
     if (atom) wc->lpszClassName = name;
 
-- 
GitLab

