From bc28fb8ede755b55fd76a595b895573ece98e3ad Mon Sep 17 00:00:00 2001
From: Paul Gofman <pgofman@codeweavers.com>
Date: Thu, 17 Jul 2025 19:08:42 -0600
Subject: [PATCH] ntdll: Factor out test_alert_with_status().

---
 dlls/ntdll/unix/server.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
index 3969c2bb898..e9d11f12e42 100644
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -851,15 +851,24 @@ NTSTATUS WINAPI NtContinueEx( CONTEXT *context, KCONTINUE_ARGUMENT *args )
 
 
 /***********************************************************************
- *              NtTestAlert  (NTDLL.@)
+ *              test_alert_with_status
  */
-NTSTATUS WINAPI NtTestAlert(void)
+static void test_alert_with_status( NTSTATUS ret_status )
 {
     struct user_apc apc;
     NTSTATUS status;
 
     status = server_select( NULL, 0, SELECT_INTERRUPTIBLE | SELECT_ALERTABLE, 0, NULL, &apc );
-    if (status == STATUS_USER_APC) invoke_user_apc( NULL, &apc, STATUS_SUCCESS );
+    if (status == STATUS_USER_APC) invoke_user_apc( NULL, &apc, ret_status );
+}
+
+
+/***********************************************************************
+ *              NtTestAlert  (NTDLL.@)
+ */
+NTSTATUS WINAPI NtTestAlert(void)
+{
+    test_alert_with_status( STATUS_SUCCESS );
     return STATUS_SUCCESS;
 }
 
-- 
GitLab

