From bd28a5ec39608eeee4e7b5d87e1e699864095510 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Fri, 29 Aug 2025 04:04:04 -0700
Subject: [PATCH] mmdevapi: Return the ceiling for min/default shared mode
 periods.

---
 dlls/mmdevapi/client.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/dlls/mmdevapi/client.c b/dlls/mmdevapi/client.c
index 07390b16527..af652470568 100644
--- a/dlls/mmdevapi/client.c
+++ b/dlls/mmdevapi/client.c
@@ -1063,10 +1063,9 @@ static HRESULT WINAPI client_GetSharedModeEnginePeriod(IAudioClient3 *iface,
     if (FAILED(hr = get_periods(This, &def_period, &min_period)))
         return hr;
 
-    *default_period_frames = def_period * format->nSamplesPerSec / (REFERENCE_TIME)10000000;
-    *min_period_frames     = (min_period * format->nSamplesPerSec + 10000000 - 1) / (REFERENCE_TIME)10000000;
-    *default_period_frames = max( *default_period_frames, *min_period_frames );
-    *max_period_frames     = *default_period_frames;
+    *default_period_frames = ((def_period * format->nSamplesPerSec) + (REFERENCE_TIME)9999999) / (REFERENCE_TIME)10000000;
+    *min_period_frames     = ((min_period * format->nSamplesPerSec) + (REFERENCE_TIME)9999999) / (REFERENCE_TIME)10000000;
+    *max_period_frames     = 512;
     *unit_period_frames    = 1;
 
     return hr;
-- 
2.50.1

