From 05c3eba273ef88690bb95a288bd03db617d8d158 Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Thu, 27 Feb 2025 09:26:03 -0800
Subject: [PATCH 08/10] winepulse: Align stream buffer to 64 bytes.

---
 dlls/winepulse.drv/pulse.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/winepulse.drv/pulse.c b/dlls/winepulse.drv/pulse.c
index 04e51739f24..37d5ed367ca 100644
--- a/dlls/winepulse.drv/pulse.c
+++ b/dlls/winepulse.drv/pulse.c
@@ -1217,12 +1217,13 @@ static NTSTATUS pulse_create_stream(void *args)
         return STATUS_SUCCESS;
     }
 
-    if (!(stream = calloc(1, sizeof(*stream))))
+    if (!(stream = aligned_alloc(64, sizeof(*stream))))
     {
         pulse_unlock();
         params->result = E_OUTOFMEMORY;
         return STATUS_SUCCESS;
     }
+    memset(stream, 0, sizeof(*stream));
 
     stream->dataflow = params->flow;
     for (i = 0; i < ARRAY_SIZE(stream->vol); ++i)
-- 
2.51.0

