From 39046007acf60453c406ebcaf9cea7636718627c Mon Sep 17 00:00:00 2001
From: William Horvath <william@horvath.blog>
Date: Sat, 4 Jan 2025 10:15:33 -0800
Subject: [PATCH] HACK: win32u: Return async keyboard state in
 NtUserGetKeyboardState.

---
 dlls/win32u/input.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index ddab781b7aa..4be3e2e5a6c 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -996,17 +996,17 @@ SHORT WINAPI NtUserGetKeyState( INT vkey )
 BOOL WINAPI NtUserGetKeyboardState( BYTE *state )
 {
     struct object_lock lock = OBJECT_LOCK_INIT;
-    const input_shm_t *input_shm;
+    const desktop_shm_t *desktop_shm;
     NTSTATUS status;
     UINT i;
 
     TRACE("(%p)\n", state);
 
-    while ((status = get_shared_input( GetCurrentThreadId(), &lock, &input_shm )) == STATUS_PENDING)
-        memcpy( state, (const void *)input_shm->keystate, 256 );
+    while ((status = get_shared_desktop( &lock, &desktop_shm )) == STATUS_PENDING)
+        memcpy( state, (const void *)desktop_shm->keystate, 256 );
     if (status) memset( state, 0, 256 );
+    else for (i = 0; i < 256; i++) state[i] &= 0x81;
 
-    for (i = 0; i < 256; i++) state[i] &= 0x81;
     return TRUE;
 }
 
-- 
2.47.1

