#https://gitlab.winehq.org/wine/wine/-/merge_requests/5896#note_73913
diff --git a/dlls/win32u/input.c b/dlls/win32u/input.c
index 4079c9ae644..1550b2fc0d7 100644
--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -813,7 +813,6 @@ SHORT WINAPI NtUserGetAsyncKeyState( INT key )
     struct object_lock lock = OBJECT_LOCK_INIT;
     NTSTATUS status;
     BYTE state = 0;
-    SHORT ret = 0;
 
     if (key < 0 || key >= 256) return 0;
 
@@ -822,23 +821,7 @@ SHORT WINAPI NtUserGetAsyncKeyState( INT key )
     while ((status = get_shared_desktop( &lock, &desktop_shm )) == STATUS_PENDING)
         state = desktop_shm->keystate[key];
 
-    if (status) return 0;
-    if (!(state & 0x40)) return (state & 0x80) << 8;
-
-    /* Need to make a server call to reset the last pressed bit */
-    SERVER_START_REQ( get_key_state )
-    {
-        req->async = 1;
-        req->key = key;
-        if (!wine_server_call( req ))
-        {
-            if (reply->state & 0x40) ret |= 0x0001;
-            if (reply->state & 0x80) ret |= 0x8000;
-        }
-    }
-    SERVER_END_REQ;
-
-    return ret;
+    return (state & 0x80) << 8;
 }
 
 /***********************************************************************
