From 5378af47b690028206027cd0983e316a10d31db3 Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Thu, 6 Mar 2025 13:47:15 +0100
Subject: [PATCH] include: Add thread priority constants.

---
 include/winnt.h | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/include/winnt.h b/include/winnt.h
index 722d2c3a542..d9f7b197da7 100644
--- a/include/winnt.h
+++ b/include/winnt.h
@@ -5358,6 +5358,11 @@ typedef enum tagSID_NAME_USE {
 #define THREAD_BASE_PRIORITY_MIN   -2
 #define THREAD_BASE_PRIORITY_IDLE  -15
 
+#define LOW_PRIORITY 0              /* Lowest thread priority level */
+#define LOW_REALTIME_PRIORITY 16    /* Lowest realtime priority level */
+#define HIGH_PRIORITY 31            /* Highest thread priority level */
+#define MAXIMUM_PRIORITY 32         /* Number of thread priority levels */
+
 typedef struct _QUOTA_LIMITS {
     SIZE_T PagedPoolLimit;
     SIZE_T NonPagedPoolLimit;
-- 
GitLab

