From 22cdc453c31df5e432cde96f4d08d59fd13d91bc Mon Sep 17 00:00:00 2001
From: Marc-Aurel Zent <mzent@codeweavers.com>
Date: Thu, 6 Mar 2025 17:24:07 +0100
Subject: [PATCH] kernelbase: Use correct priority in GetThreadPriority.

---
 dlls/kernelbase/thread.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/kernelbase/thread.c b/dlls/kernelbase/thread.c
index 82096fd45f3..82ade671d1a 100644
--- a/dlls/kernelbase/thread.c
+++ b/dlls/kernelbase/thread.c
@@ -274,7 +274,7 @@ INT WINAPI DECLSPEC_HOTPATCH GetThreadPriority( HANDLE thread )
     if (!set_ntstatus( NtQueryInformationThread( thread, ThreadBasicInformation,
                                                  &info, sizeof(info), NULL )))
         return THREAD_PRIORITY_ERROR_RETURN;
-    return info.Priority;
+    return info.BasePriority;
 }
 
 
-- 
GitLab

