From 94b7919d58b5ae440d91a9004b0c67ec1c49bae5 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Sat, 22 Jun 2024 11:53:55 +0200
Subject: [PATCH] include: Don't import atexit from CRTs.

---
 include/msvcrt/stdlib.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/msvcrt/stdlib.h b/include/msvcrt/stdlib.h
index f3df662528a..df307196a12 100644
--- a/include/msvcrt/stdlib.h
+++ b/include/msvcrt/stdlib.h
@@ -203,7 +203,7 @@ _ACRTIMP DECLSPEC_NORETURN void __cdecl _Exit(int);
 _ACRTIMP DECLSPEC_NORETURN void __cdecl _exit(int);
 _ACRTIMP DECLSPEC_NORETURN void __cdecl abort(void);
 _ACRTIMP int           __cdecl abs(int);
-_ACRTIMP int           __cdecl atexit(void (__cdecl *)(void));
+extern int             __cdecl atexit(void (__cdecl *)(void));
 _ACRTIMP double        __cdecl atof(const char*);
 _ACRTIMP int           __cdecl atoi(const char*);
 _ACRTIMP int           __cdecl _atoi_l(const char*,_locale_t);
-- 
GitLab

