From 879b92d01940d138b133bed29e1635f1e602bb1c Mon Sep 17 00:00:00 2001
From: Matteo Bruni <mbruni@codeweavers.com>
Date: Fri, 11 Apr 2025 16:25:34 +0200
Subject: [PATCH] ntdll: Factor out a cancel_io() function.

---
 dlls/ntdll/unix/file.c | 38 ++++++++++++++++----------------------
 1 file changed, 16 insertions(+), 22 deletions(-)

diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index a475fcc3225..aacd96f5a3d 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -6421,19 +6421,16 @@ NTSTATUS WINAPI NtFlushBuffersFileEx( HANDLE handle, ULONG flags, void *params,
 }
 
 
-/**************************************************************************
- *           NtCancelIoFile   (NTDLL.@)
- */
-NTSTATUS WINAPI NtCancelIoFile( HANDLE handle, IO_STATUS_BLOCK *io_status )
+static NTSTATUS cancel_io( HANDLE handle, IO_STATUS_BLOCK *io, IO_STATUS_BLOCK *io_status,
+                           BOOL only_thread )
 {
     unsigned int status;
 
-    TRACE( "%p %p\n", handle, io_status );
-
     SERVER_START_REQ( cancel_async )
     {
         req->handle      = wine_server_obj_handle( handle );
-        req->only_thread = TRUE;
+        req->iosb        = wine_server_client_ptr( io );
+        req->only_thread = only_thread;
         if (!(status = wine_server_call( req )))
         {
             io_status->Status = status;
@@ -6446,28 +6443,25 @@ NTSTATUS WINAPI NtCancelIoFile( HANDLE handle, IO_STATUS_BLOCK *io_status )
 }
 
 
+/**************************************************************************
+ *           NtCancelIoFile   (NTDLL.@)
+ */
+NTSTATUS WINAPI NtCancelIoFile( HANDLE handle, IO_STATUS_BLOCK *io_status )
+{
+    TRACE( "%p %p\n", handle, io_status );
+
+    return cancel_io( handle, NULL, io_status, TRUE );
+}
+
+
 /**************************************************************************
  *           NtCancelIoFileEx   (NTDLL.@)
  */
 NTSTATUS WINAPI NtCancelIoFileEx( HANDLE handle, IO_STATUS_BLOCK *io, IO_STATUS_BLOCK *io_status )
 {
-    unsigned int status;
-
     TRACE( "%p %p %p\n", handle, io, io_status );
 
-    SERVER_START_REQ( cancel_async )
-    {
-        req->handle = wine_server_obj_handle( handle );
-        req->iosb   = wine_server_client_ptr( io );
-        if (!(status = wine_server_call( req )))
-        {
-            io_status->Status = status;
-            io_status->Information = 0;
-        }
-    }
-    SERVER_END_REQ;
-
-    return status;
+    return cancel_io( handle, io, io_status, FALSE );
 }
 
 
-- 
GitLab

