From a190b1daacb75acbd987bf6939c89be4b29a8ade Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Mon, 5 May 2025 11:03:14 +0200
Subject: [PATCH] server: Introduce a find_async_from_user helper.

---
 server/async.c | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/server/async.c b/server/async.c
index 4068f744567..1602746a969 100644
--- a/server/async.c
+++ b/server/async.c
@@ -576,6 +576,16 @@ int async_waiting( struct async_queue *queue )
     return !async->terminated;
 }
 
+static struct async *find_async_from_user( struct process *process, client_ptr_t user )
+{
+    struct async *async;
+
+    LIST_FOR_EACH_ENTRY( async, &process->asyncs, struct async, process_entry )
+        if (async->data.user == user) return async;
+
+    return NULL;
+}
+
 static int cancel_async( struct process *process, struct object *obj, struct thread *thread, client_ptr_t iosb )
 {
     struct async *async;
@@ -830,17 +840,10 @@ DECL_HANDLER(cancel_async)
 /* get async result from associated iosb */
 DECL_HANDLER(get_async_result)
 {
-    struct iosb *iosb = NULL;
     struct async *async;
+    struct iosb *iosb;
 
-    LIST_FOR_EACH_ENTRY( async, &current->process->asyncs, struct async, process_entry )
-        if (async->data.user == req->user_arg)
-        {
-            iosb = async->iosb;
-            break;
-        }
-
-    if (!iosb)
+    if (!(async = find_async_from_user( current->process, req->user_arg )) || !(iosb = async->iosb))
     {
         set_error( STATUS_INVALID_PARAMETER );
         return;
-- 
GitLab

