From 15fbaa0a95044fe6e49b4cd81f85b6b1918d8784 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 12 Sep 2025 17:43:22 +0200
Subject: [PATCH] win32u: Remove window surface flush in expose_window_surface.

---
 dlls/win32u/window.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
index bb7718346c8..73d9fe283e0 100644
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -2291,7 +2291,6 @@ static BOOL expose_window_surface( HWND hwnd, UINT flags, const RECT *rect, UINT
         add_bounds_rect( &surface->bounds, &exposed_rect );
     }
     window_surface_unlock( surface );
-    if (surface->alpha_mask) window_surface_flush( surface );
     window_surface_release( surface );
     return TRUE;
 }
-- 
GitLab

