From 6be9667edf47f6cc882b594d762c9de06a254572 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Mon, 29 Jul 2024 10:41:23 +0200
Subject: [PATCH 2/4] mf/sar: Allow starting audio renderer again after pausing
 it.

---
 dlls/mf/sar.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/dlls/mf/sar.c b/dlls/mf/sar.c
index 84824f954dd..4adbc28a39d 100644
--- a/dlls/mf/sar.c
+++ b/dlls/mf/sar.c
@@ -636,7 +636,7 @@ static HRESULT WINAPI audio_renderer_clock_sink_OnClockStart(IMFClockStateSink *
     EnterCriticalSection(&renderer->cs);
     if (renderer->audio_client)
     {
-        if (renderer->state == STREAM_STATE_STOPPED)
+        if (renderer->state == STREAM_STATE_PAUSED || renderer->state == STREAM_STATE_STOPPED)
         {
             if (FAILED(hr = IAudioClient_Start(renderer->audio_client)))
                 WARN("Failed to start audio client, hr %#lx.\n", hr);
-- 
2.46.0

