From d943263c3e3b14882ad8671100f4ffd1e6eed48b Mon Sep 17 00:00:00 2001
From: Bartosz Kosiorek <gang65@poczta.onet.pl>
Date: Sat, 15 Mar 2025 12:53:33 +0100
Subject: [PATCH] gdiplus/test: Add additional test for GdipAddPathArc

---
 dlls/gdiplus/tests/graphicspath.c | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/dlls/gdiplus/tests/graphicspath.c b/dlls/gdiplus/tests/graphicspath.c
index a0fcbde4a0a..28db22227b5 100644
--- a/dlls/gdiplus/tests/graphicspath.c
+++ b/dlls/gdiplus/tests/graphicspath.c
@@ -1253,7 +1253,14 @@ static path_test_t flattenline_path[] = {
     {50.0, 100.0, PathPointTypeLine,  0, 0} /*1*/
     };
 
-static path_test_t flattenarc_path[] = {
+static path_test_t arc_path3[] = {
+    {100.0, 25.0, PathPointTypeStart,  0, 0}, /*0*/
+    {100.0, 38.8, PathPointTypeBezier, 0, 0}, /*1*/
+    {77.6,  50.0, PathPointTypeBezier, 0, 0}, /*2*/
+    {50.0,  50.0, PathPointTypeBezier, 0, 0}  /*3*/
+    };
+
+static path_test_t flattenarc_path3[] = {
     {100.0, 25.0,PathPointTypeStart, 0, 0}, /*0*/
     {99.0, 30.0, PathPointTypeLine,  0, 0}, /*1*/
     {96.0, 34.8, PathPointTypeLine,  0, 0}, /*2*/
@@ -1318,9 +1325,11 @@ static void test_flatten(void)
     expect(Ok, status);
     status = GdipAddPathArc(path, 0.0, 0.0, 100.0, 50.0, 0.0, 90.0);
     expect(Ok, status);
+    ok_path(path, arc_path3, ARRAY_SIZE(arc_path3), FALSE);
+
     status = GdipFlattenPath(path, NULL, 1.0);
     expect(Ok, status);
-    ok_path(path, flattenarc_path, ARRAY_SIZE(flattenarc_path), TRUE);
+    ok_path(path, flattenarc_path3, ARRAY_SIZE(flattenarc_path3), TRUE);
 
     /* easy case - quater of a full circle */
     status = GdipResetPath(path);
-- 
GitLab

