From cfc3cd1dee577c5551e82821e1b1b65036e93ae3 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Thu, 8 Aug 2024 15:01:23 +0200
Subject: [PATCH 4/5] winex11: Use the default / dummy surface with foreign
 windows.

---
 dlls/winex11.drv/bitblt.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/dlls/winex11.drv/bitblt.c b/dlls/winex11.drv/bitblt.c
index 7a3785ae398..7c5d0643570 100644
--- a/dlls/winex11.drv/bitblt.c
+++ b/dlls/winex11.drv/bitblt.c
@@ -1994,6 +1994,7 @@ BOOL X11DRV_CreateWindowSurface( HWND hwnd, BOOL layered, const RECT *surface_re
     TRACE( "hwnd %p, layered %u, surface_rect %s, surface %p\n", hwnd, layered, wine_dbgstr_rect( surface_rect ), surface );
 
     if (!(data = get_win_data( hwnd ))) return TRUE; /* use default surface */
+    if (!data->whole_window) goto done; /* use the dummy/offscreen surfaces for foreign windows */
     if ((previous = *surface) && previous->funcs == &x11drv_surface_funcs)
     {
         if (data->whole_window == get_x11_surface(previous)->window) goto done; /* use default surface */
-- 
2.46.0

