From e7a63da1a6a3328747f3be4004fefd44d6d3712a Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Wed, 26 Jan 2022 00:25:51 +0900
Subject: [PATCH 41/51] loader: Enable dumping additional vectors in
 dump_auxiliary.

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---
 loader/preloader.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/loader/preloader.c b/loader/preloader.c
index 12a98745b16..d8dd210e50a 100644
--- a/loader/preloader.c
+++ b/loader/preloader.c
@@ -1191,6 +1191,11 @@ static void dump_auxiliary( struct wld_auxv *av )
         NAME(AT_SYSINFO),
         NAME(AT_SYSINFO_EHDR),
         NAME(AT_UID),
+        NAME(AT_SECURE),
+        NAME(AT_RANDOM),
+        NAME(AT_HWCAP2),
+        NAME(AT_EXECFN),
+        NAME(AT_MINSIGSTKSZ),
         { 0, NULL }
     };
 #undef NAME
-- 
2.53.0

