From ee9f60dcff034db3e203bd6a6ca90f1f7c3c9971 Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Sat, 5 Mar 2022 23:55:01 +0900
Subject: [PATCH 42/51] loader: Enable all remap logic by default.

Signed-off-by: Jinoh Kang <jinoh.kang.kr@gmail.com>
---
 loader/preloader.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/loader/preloader.c b/loader/preloader.c
index d8dd210e50a..3348918a2d2 100644
--- a/loader/preloader.c
+++ b/loader/preloader.c
@@ -293,11 +293,11 @@ enum remap_policy
     REMAP_POLICY_SKIP = 2,
     LAST_REMAP_POLICY,
 
-    REMAP_POLICY_DEFAULT_VDSO    = REMAP_POLICY_SKIP,
+    REMAP_POLICY_DEFAULT_VDSO    = REMAP_POLICY_ON_CONFLICT,
 #ifdef __arm__
-    REMAP_POLICY_DEFAULT_SIGPAGE = REMAP_POLICY_SKIP,
+    REMAP_POLICY_DEFAULT_SIGPAGE = REMAP_POLICY_ON_CONFLICT,
 #endif
-    REMAP_POLICY_DEFAULT_STACK   = REMAP_POLICY_SKIP,
+    REMAP_POLICY_DEFAULT_STACK   = REMAP_POLICY_ON_CONFLICT,
 };
 
 /*
-- 
2.53.0

