From b9879d5adc1cd98a6ceb7d087c4533fdd487bedb Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Bernon?= <rbernon@codeweavers.com>
Date: Fri, 9 Aug 2024 13:36:26 +0200
Subject: [PATCH] winewayland: Remove now unnecessary WindowPosChanging checks.

---
 dlls/winewayland.drv/window.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git b/dlls/winewayland.drv/window.c a/dlls/winewayland.drv/window.c
index 7c694246fb7..dc6a6b7e03a 100644
--- b/dlls/winewayland.drv/window.c
+++ a/dlls/winewayland.drv/window.c
@@ -431,15 +431,23 @@ BOOL WAYLAND_WindowPosChanging(HWND hwnd, UINT swp_flags, BOOL shaped, const REC
                                const RECT *client_rect, RECT *visible_rect)
 {
     struct wayland_win_data *data = wayland_win_data_get(hwnd);
+    HWND parent;
+    BOOL ret = FALSE;
 
     TRACE("hwnd %p, swp_flags %04x, shaped %u, window_rect %s, client_rect %s, visible_rect %s\n",
           hwnd, swp_flags, shaped, wine_dbgstr_rect(window_rect), wine_dbgstr_rect(client_rect),
           wine_dbgstr_rect(visible_rect));
 
     if (!data && !(data = wayland_win_data_create(hwnd, window_rect, client_rect))) return FALSE; /* use default surface */
-    wayland_win_data_release(data);
 
-    return TRUE;
+    parent = NtUserGetAncestor(hwnd, GA_PARENT);
+    if ((parent && parent != NtUserGetDesktopWindow())) goto done; /* use default surface */
+
+    ret = TRUE;
+
+done:
+    wayland_win_data_release(data);
+    return ret;
 }
 
 
-- 
2.46.0

